{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build distance maps from antigenic escape assay data\n",
    "\n",
    "[Greaney et al. 2021](https://doi.org/10.1101/2020.12.31.425021) identified positions in SARS-CoV-2 spike (S) that allowed specific viruses to bind yeast-displayed ACE2 in the presence of antibodies from human polyclonal sera. These assays quantify the degree to which specific mutations enabled antigenic escape in a measure called the `mut_escape`. The resulting data contain the `mut_escape` value for each amino acid mutation from wildtype at each receptor-binding domain (RBD) position in S and specific human serum. The total escape at each site is given by the `total_escape` column.\n",
    "\n",
    "To understand whether we can use these data to track antigenic evolution of SARS-CoV-2, we developed the following distance maps for use in Nextstrain workflows and eventual visualization in phylogenetic trees:\n",
    "\n",
    "1. antigenic escape epitope sites: a binary distance map of positions with the strongest immune escape values across all available human sera. This map reflects putative epitope sites akin to the Koel et al. 2013 sites without any consideration of what mutations occur at those sites. This map is most likely to avoid overfitting to a specific serum sample.\n",
    "\n",
    "2. antigenic escape weighted epitope sites: a floating point distance map for each position in S reflecting the proportion of total antigenic escape provided by that position (total `mut_escape` at that site divided by the total `mut_escape` across all sites and averaged across all available sera). This map avoids using arbitrary thresholds of `mut_escape` values to identify epitope sites by weighting all sites across all sera.\n",
    "\n",
    "3. antigenic escape by mutation: a site- and mutation-specific floating point distance map reflecting the specific mutational effect of a specific derived state from the corresponding wildtype state. This map provides distances based directly on the absolute measurements of the mutation escape at each position.\n",
    "\n",
    "Below, we build each map in order of increasing complexity and export the corresponding maps in the format required by the `augur distance` command."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports and configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_url = \"https://raw.githubusercontent.com/jbloomlab/SARS-CoV-2-RBD_MAP_HAARVI_sera/main/results/supp_data/human_sera_6m0j_dms-view_data.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "antigenic_escape_distance_map = \"defaults/distance_maps/greaney_2021_antigenic_escape_ep.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "weighted_antigenic_escape_distance_map = \"defaults/distance_maps/greaney_2021_weighted_antigenic_escape_ep.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_antigenic_escape_distance_map = \"defaults/distance_maps/greaney_2021_mutation_antigenic_escape.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_conditions = [\n",
    "    'subject A (day 120)',\n",
    "    'subject B (day 113)',\n",
    "    'subject C (day 104)',\n",
    "    'subject D (day 76)',\n",
    "    'subject E (day 104)',\n",
    "    'subject F (day 115)',\n",
    "    'subject G (day 94)',\n",
    "    'subject H (day 152)',\n",
    "    'subject I (day 102)',\n",
    "    'subject J (day 121)',\n",
    "    'subject K (day 103)'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(data_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>site</th>\n",
       "      <th>label_site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mutation</th>\n",
       "      <th>protein_chain</th>\n",
       "      <th>protein_site</th>\n",
       "      <th>mut_escape color ACE2 bind</th>\n",
       "      <th>site_total escape</th>\n",
       "      <th>site_max escape</th>\n",
       "      <th>color_for_mutation</th>\n",
       "      <th>mut_escape color RBD expr</th>\n",
       "      <th>mut_escape color gray</th>\n",
       "      <th>mut_escape color func group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.002020</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#692505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.005616</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#662505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.002535</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#662505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.003032</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#722805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.003113</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#6a2605</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             condition  site  label_site wildtype mutation protein_chain  \\\n",
       "0  subject H (day 152)     1         331        N        A             E   \n",
       "1  subject H (day 152)     1         331        N        D             E   \n",
       "2  subject H (day 152)     1         331        N        E             E   \n",
       "3  subject H (day 152)     1         331        N        F             E   \n",
       "4  subject H (day 152)     1         331        N        G             E   \n",
       "\n",
       "   protein_site  mut_escape color ACE2 bind  site_total escape  \\\n",
       "0           331                    0.002020            0.04926   \n",
       "1           331                    0.005616            0.04926   \n",
       "2           331                    0.002535            0.04926   \n",
       "3           331                    0.003032            0.04926   \n",
       "4           331                    0.003113            0.04926   \n",
       "\n",
       "   site_max escape color_for_mutation  mut_escape color RBD expr  \\\n",
       "0         0.007632            #692505                        NaN   \n",
       "1         0.007632            #662505                        NaN   \n",
       "2         0.007632            #662505                        NaN   \n",
       "3         0.007632            #722805                        NaN   \n",
       "4         0.007632            #6a2605                        NaN   \n",
       "\n",
       "   mut_escape color gray  mut_escape color func group  \n",
       "0                    NaN                          NaN  \n",
       "1                    NaN                          NaN  \n",
       "2                    NaN                          NaN  \n",
       "3                    NaN                          NaN  \n",
       "4                    NaN                          NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(23,)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"condition\"].value_counts().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df[\"condition\"].isin(valid_conditions)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_subject_from_condition(condition):\n",
    "    return condition.split(\"(\")[0].strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_day_from_condition(condition):\n",
    "    return int(condition.split(\" \")[-1].replace(\")\", \"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"subject\"] = df[\"condition\"].apply(get_subject_from_condition)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"day\"] = df[\"condition\"].apply(get_day_from_condition)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "subject G    7792\n",
       "subject H    7784\n",
       "subject D    7780\n",
       "subject K    7780\n",
       "subject E    7780\n",
       "subject C    7780\n",
       "subject A    7772\n",
       "subject B    7772\n",
       "subject J    7772\n",
       "subject I    7772\n",
       "subject F    7772\n",
       "Name: subject, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"subject\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "104    15560\n",
       "94      7792\n",
       "152     7784\n",
       "103     7780\n",
       "76      7780\n",
       "121     7772\n",
       "120     7772\n",
       "115     7772\n",
       "113     7772\n",
       "102     7772\n",
       "Name: day, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"day\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['subject A (day 120)',\n",
       " 'subject B (day 113)',\n",
       " 'subject C (day 104)',\n",
       " 'subject D (day 76)',\n",
       " 'subject E (day 104)',\n",
       " 'subject F (day 115)',\n",
       " 'subject G (day 94)',\n",
       " 'subject H (day 152)',\n",
       " 'subject I (day 102)',\n",
       " 'subject J (day 121)',\n",
       " 'subject K (day 103)']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(df.sort_values(\n",
    "    [\"subject\", \"day\"],\n",
    "    ascending=False\n",
    ").groupby(\n",
    "    [\"subject\"],\n",
    "    sort=False\n",
    ").first()[\"condition\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>site</th>\n",
       "      <th>label_site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mutation</th>\n",
       "      <th>protein_chain</th>\n",
       "      <th>protein_site</th>\n",
       "      <th>mut_escape color ACE2 bind</th>\n",
       "      <th>site_total escape</th>\n",
       "      <th>site_max escape</th>\n",
       "      <th>color_for_mutation</th>\n",
       "      <th>mut_escape color RBD expr</th>\n",
       "      <th>mut_escape color gray</th>\n",
       "      <th>mut_escape color func group</th>\n",
       "      <th>subject</th>\n",
       "      <th>day</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.002020</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#692505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>subject H</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.005616</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#662505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>subject H</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.002535</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#662505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>subject H</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.003032</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#722805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>subject H</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.003113</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#6a2605</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>subject H</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             condition  site  label_site wildtype mutation protein_chain  \\\n",
       "0  subject H (day 152)     1         331        N        A             E   \n",
       "1  subject H (day 152)     1         331        N        D             E   \n",
       "2  subject H (day 152)     1         331        N        E             E   \n",
       "3  subject H (day 152)     1         331        N        F             E   \n",
       "4  subject H (day 152)     1         331        N        G             E   \n",
       "\n",
       "   protein_site  mut_escape color ACE2 bind  site_total escape  \\\n",
       "0           331                    0.002020            0.04926   \n",
       "1           331                    0.005616            0.04926   \n",
       "2           331                    0.002535            0.04926   \n",
       "3           331                    0.003032            0.04926   \n",
       "4           331                    0.003113            0.04926   \n",
       "\n",
       "   site_max escape color_for_mutation  mut_escape color RBD expr  \\\n",
       "0         0.007632            #692505                        NaN   \n",
       "1         0.007632            #662505                        NaN   \n",
       "2         0.007632            #662505                        NaN   \n",
       "3         0.007632            #722805                        NaN   \n",
       "4         0.007632            #6a2605                        NaN   \n",
       "\n",
       "   mut_escape color gray  mut_escape color func group    subject  day  \n",
       "0                    NaN                          NaN  subject H  152  \n",
       "1                    NaN                          NaN  subject H  152  \n",
       "2                    NaN                          NaN  subject H  152  \n",
       "3                    NaN                          NaN  subject H  152  \n",
       "4                    NaN                          NaN  subject H  152  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([331, 332, 333, 334, 335, 336, 337, 338, 339, 340, 341, 342, 344,\n",
       "       345, 346, 347, 348, 349, 351, 352, 353, 354, 356, 357, 358, 359,\n",
       "       360, 361, 362, 363, 365, 366, 367, 368, 369, 370, 371, 372, 373,\n",
       "       374, 375, 376, 377, 378, 380, 381, 382, 383, 384, 385, 386, 387,\n",
       "       388, 389, 390, 391, 392, 393, 394, 395, 396, 397, 399, 401, 402,\n",
       "       403, 404, 405, 406, 407, 408, 409, 410, 411, 413, 414, 415, 417,\n",
       "       418, 419, 420, 421, 424, 425, 427, 428, 429, 430, 433, 434, 435,\n",
       "       437, 438, 439, 440, 441, 443, 444, 445, 446, 447, 448, 449, 450,\n",
       "       451, 452, 453, 455, 456, 458, 459, 460, 461, 462, 463, 464, 465,\n",
       "       466, 468, 469, 470, 471, 472, 473, 474, 475, 476, 477, 478, 479,\n",
       "       481, 482, 483, 484, 485, 486, 487, 489, 490, 492, 493, 494, 496,\n",
       "       498, 499, 500, 501, 503, 504, 505, 506, 507, 508, 510, 512, 513,\n",
       "       514, 515, 516, 517, 518, 519, 520, 521, 522, 523, 524, 525, 526,\n",
       "       527, 528, 529, 530, 531])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"label_site\"].drop_duplicates().sort_values().values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# All sites in this analysis are in the same gene, spike (S).\n",
    "df[\"gene\"] = \"S\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"gene_site\"] = df[\"label_site\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>site</th>\n",
       "      <th>label_site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mutation</th>\n",
       "      <th>protein_chain</th>\n",
       "      <th>protein_site</th>\n",
       "      <th>mut_escape color ACE2 bind</th>\n",
       "      <th>site_total escape</th>\n",
       "      <th>site_max escape</th>\n",
       "      <th>color_for_mutation</th>\n",
       "      <th>mut_escape color RBD expr</th>\n",
       "      <th>mut_escape color gray</th>\n",
       "      <th>mut_escape color func group</th>\n",
       "      <th>subject</th>\n",
       "      <th>day</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.002020</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#692505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>subject H</td>\n",
       "      <td>152</td>\n",
       "      <td>S</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.005616</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#662505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>subject H</td>\n",
       "      <td>152</td>\n",
       "      <td>S</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.002535</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#662505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>subject H</td>\n",
       "      <td>152</td>\n",
       "      <td>S</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.003032</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#722805</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>subject H</td>\n",
       "      <td>152</td>\n",
       "      <td>S</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>1</td>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>E</td>\n",
       "      <td>331</td>\n",
       "      <td>0.003113</td>\n",
       "      <td>0.04926</td>\n",
       "      <td>0.007632</td>\n",
       "      <td>#6a2605</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>subject H</td>\n",
       "      <td>152</td>\n",
       "      <td>S</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             condition  site  label_site wildtype mutation protein_chain  \\\n",
       "0  subject H (day 152)     1         331        N        A             E   \n",
       "1  subject H (day 152)     1         331        N        D             E   \n",
       "2  subject H (day 152)     1         331        N        E             E   \n",
       "3  subject H (day 152)     1         331        N        F             E   \n",
       "4  subject H (day 152)     1         331        N        G             E   \n",
       "\n",
       "   protein_site  mut_escape color ACE2 bind  site_total escape  \\\n",
       "0           331                    0.002020            0.04926   \n",
       "1           331                    0.005616            0.04926   \n",
       "2           331                    0.002535            0.04926   \n",
       "3           331                    0.003032            0.04926   \n",
       "4           331                    0.003113            0.04926   \n",
       "\n",
       "   site_max escape color_for_mutation  mut_escape color RBD expr  \\\n",
       "0         0.007632            #692505                        NaN   \n",
       "1         0.007632            #662505                        NaN   \n",
       "2         0.007632            #662505                        NaN   \n",
       "3         0.007632            #722805                        NaN   \n",
       "4         0.007632            #6a2605                        NaN   \n",
       "\n",
       "   mut_escape color gray  mut_escape color func group    subject  day gene  \\\n",
       "0                    NaN                          NaN  subject H  152    S   \n",
       "1                    NaN                          NaN  subject H  152    S   \n",
       "2                    NaN                          NaN  subject H  152    S   \n",
       "3                    NaN                          NaN  subject H  152    S   \n",
       "4                    NaN                          NaN  subject H  152    S   \n",
       "\n",
       "   gene_site  \n",
       "0        331  \n",
       "1        331  \n",
       "2        331  \n",
       "3        331  \n",
       "4        331  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_per_site = df.loc[:, [\"label_site\", \"wildtype\", \"mutation\"]].drop_duplicates().copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label_site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    label_site wildtype mutation\n",
       "0          331        N        A\n",
       "1          331        N        D\n",
       "2          331        N        E\n",
       "3          331        N        F\n",
       "4          331        N        G\n",
       "5          331        N        H\n",
       "6          331        N        I\n",
       "7          331        N        K\n",
       "8          331        N        L\n",
       "9          331        N        M\n",
       "10         331        N        P\n",
       "11         331        N        Q\n",
       "12         331        N        R\n",
       "13         331        N        S\n",
       "14         331        N        T\n",
       "15         331        N        V"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_per_site.query(\"label_site == 331\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20,)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_per_site[\"mutation\"].drop_duplicates().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "subject G (day 94)     7792\n",
       "subject H (day 152)    7784\n",
       "subject K (day 103)    7780\n",
       "subject D (day 76)     7780\n",
       "subject C (day 104)    7780\n",
       "subject E (day 104)    7780\n",
       "subject F (day 115)    7772\n",
       "subject B (day 113)    7772\n",
       "subject A (day 120)    7772\n",
       "subject I (day 102)    7772\n",
       "subject J (day 121)    7772\n",
       "Name: condition, dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"condition\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "subject G    7792\n",
       "subject H    7784\n",
       "subject D    7780\n",
       "subject K    7780\n",
       "subject E    7780\n",
       "subject C    7780\n",
       "subject A    7772\n",
       "subject B    7772\n",
       "subject J    7772\n",
       "subject I    7772\n",
       "subject F    7772\n",
       "Name: subject, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"subject\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identify antigenic escape epitope sites\n",
    "\n",
    "Calculate site-specific total mut_escape values, plot their distributions per serum, and identify those positions with values greater than 4 standard deviations from the mean by serum."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "escape_column = \"site_total escape\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "site_columns = [\n",
    "    \"condition\",\n",
    "    \"site\",\n",
    "    \"gene\",\n",
    "    \"gene_site\",\n",
    "    escape_column\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "site_specific_df = df.loc[:, site_columns].drop_duplicates().sort_values([\"condition\", \"site\"]).reset_index().drop(columns=[\"index\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>site</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_site</th>\n",
       "      <th>site_total escape</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>1</td>\n",
       "      <td>S</td>\n",
       "      <td>331</td>\n",
       "      <td>0.1205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>2</td>\n",
       "      <td>S</td>\n",
       "      <td>332</td>\n",
       "      <td>0.6332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>3</td>\n",
       "      <td>S</td>\n",
       "      <td>333</td>\n",
       "      <td>0.1667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>4</td>\n",
       "      <td>S</td>\n",
       "      <td>334</td>\n",
       "      <td>0.5097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>5</td>\n",
       "      <td>S</td>\n",
       "      <td>335</td>\n",
       "      <td>0.6866</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             condition  site gene  gene_site  site_total escape\n",
       "0  subject A (day 120)     1    S        331             0.1205\n",
       "1  subject A (day 120)     2    S        332             0.6332\n",
       "2  subject A (day 120)     3    S        333             0.1667\n",
       "3  subject A (day 120)     4    S        334             0.5097\n",
       "4  subject A (day 120)     5    S        335             0.6866"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "site_specific_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1914, 5)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "site_specific_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "condition_specific_distributions = site_specific_df.groupby(\"condition\").aggregate({escape_column: [\"mean\", \"std\"]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "condition_specific_distributions[\"upper_threshold\"] = (\n",
    "    condition_specific_distributions[escape_column, \"mean\"] + 4 * condition_specific_distributions[escape_column, \"std\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "upper_thresholds = condition_specific_distributions.reset_index().loc[:, [\"condition\", \"upper_threshold\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "upper_thresholds.columns = upper_thresholds.columns.droplevel(level=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>upper_threshold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>2.785092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>subject B (day 113)</td>\n",
       "      <td>2.134808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>subject C (day 104)</td>\n",
       "      <td>0.305421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>subject D (day 76)</td>\n",
       "      <td>0.047514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>subject E (day 104)</td>\n",
       "      <td>3.191500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>subject F (day 115)</td>\n",
       "      <td>3.365717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>subject G (day 94)</td>\n",
       "      <td>2.024142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>2.224005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>subject I (day 102)</td>\n",
       "      <td>4.384038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>subject J (day 121)</td>\n",
       "      <td>2.049182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>subject K (day 103)</td>\n",
       "      <td>0.305892</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              condition  upper_threshold\n",
       "0   subject A (day 120)         2.785092\n",
       "1   subject B (day 113)         2.134808\n",
       "2   subject C (day 104)         0.305421\n",
       "3    subject D (day 76)         0.047514\n",
       "4   subject E (day 104)         3.191500\n",
       "5   subject F (day 115)         3.365717\n",
       "6    subject G (day 94)         2.024142\n",
       "7   subject H (day 152)         2.224005\n",
       "8   subject I (day 102)         4.384038\n",
       "9   subject J (day 121)         2.049182\n",
       "10  subject K (day 103)         0.305892"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "upper_thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "site_specific_df = site_specific_df.merge(\n",
    "    upper_thresholds,\n",
    "    on=\"condition\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>site</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_site</th>\n",
       "      <th>site_total escape</th>\n",
       "      <th>upper_threshold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>126</td>\n",
       "      <td>S</td>\n",
       "      <td>456</td>\n",
       "      <td>3.7920</td>\n",
       "      <td>2.785092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>154</td>\n",
       "      <td>S</td>\n",
       "      <td>484</td>\n",
       "      <td>3.0420</td>\n",
       "      <td>2.785092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>282</th>\n",
       "      <td>subject B (day 113)</td>\n",
       "      <td>126</td>\n",
       "      <td>S</td>\n",
       "      <td>456</td>\n",
       "      <td>3.5100</td>\n",
       "      <td>2.134808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>subject B (day 113)</td>\n",
       "      <td>154</td>\n",
       "      <td>S</td>\n",
       "      <td>484</td>\n",
       "      <td>3.0190</td>\n",
       "      <td>2.134808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>subject C (day 104)</td>\n",
       "      <td>126</td>\n",
       "      <td>S</td>\n",
       "      <td>456</td>\n",
       "      <td>0.6150</td>\n",
       "      <td>0.305421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>subject C (day 104)</td>\n",
       "      <td>154</td>\n",
       "      <td>S</td>\n",
       "      <td>484</td>\n",
       "      <td>0.7445</td>\n",
       "      <td>0.305421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>subject D (day 76)</td>\n",
       "      <td>126</td>\n",
       "      <td>S</td>\n",
       "      <td>456</td>\n",
       "      <td>0.1029</td>\n",
       "      <td>0.047514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>804</th>\n",
       "      <td>subject E (day 104)</td>\n",
       "      <td>126</td>\n",
       "      <td>S</td>\n",
       "      <td>456</td>\n",
       "      <td>3.7300</td>\n",
       "      <td>3.191500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>subject E (day 104)</td>\n",
       "      <td>142</td>\n",
       "      <td>S</td>\n",
       "      <td>472</td>\n",
       "      <td>3.6470</td>\n",
       "      <td>3.191500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>829</th>\n",
       "      <td>subject E (day 104)</td>\n",
       "      <td>154</td>\n",
       "      <td>S</td>\n",
       "      <td>484</td>\n",
       "      <td>8.5360</td>\n",
       "      <td>3.191500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>subject F (day 115)</td>\n",
       "      <td>119</td>\n",
       "      <td>S</td>\n",
       "      <td>449</td>\n",
       "      <td>3.6630</td>\n",
       "      <td>3.365717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978</th>\n",
       "      <td>subject F (day 115)</td>\n",
       "      <td>126</td>\n",
       "      <td>S</td>\n",
       "      <td>456</td>\n",
       "      <td>5.6520</td>\n",
       "      <td>3.365717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1003</th>\n",
       "      <td>subject F (day 115)</td>\n",
       "      <td>154</td>\n",
       "      <td>S</td>\n",
       "      <td>484</td>\n",
       "      <td>4.9770</td>\n",
       "      <td>3.365717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1318</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>117</td>\n",
       "      <td>S</td>\n",
       "      <td>447</td>\n",
       "      <td>2.2260</td>\n",
       "      <td>2.224005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1320</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>119</td>\n",
       "      <td>S</td>\n",
       "      <td>449</td>\n",
       "      <td>2.8720</td>\n",
       "      <td>2.224005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1323</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>122</td>\n",
       "      <td>S</td>\n",
       "      <td>452</td>\n",
       "      <td>2.5470</td>\n",
       "      <td>2.224005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1351</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>154</td>\n",
       "      <td>S</td>\n",
       "      <td>484</td>\n",
       "      <td>2.9980</td>\n",
       "      <td>2.224005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1500</th>\n",
       "      <td>subject I (day 102)</td>\n",
       "      <td>126</td>\n",
       "      <td>S</td>\n",
       "      <td>456</td>\n",
       "      <td>4.7590</td>\n",
       "      <td>4.384038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>subject I (day 102)</td>\n",
       "      <td>154</td>\n",
       "      <td>S</td>\n",
       "      <td>484</td>\n",
       "      <td>7.3300</td>\n",
       "      <td>4.384038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1527</th>\n",
       "      <td>subject I (day 102)</td>\n",
       "      <td>156</td>\n",
       "      <td>S</td>\n",
       "      <td>486</td>\n",
       "      <td>5.0250</td>\n",
       "      <td>4.384038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1530</th>\n",
       "      <td>subject I (day 102)</td>\n",
       "      <td>160</td>\n",
       "      <td>S</td>\n",
       "      <td>490</td>\n",
       "      <td>4.9600</td>\n",
       "      <td>4.384038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1848</th>\n",
       "      <td>subject K (day 103)</td>\n",
       "      <td>126</td>\n",
       "      <td>S</td>\n",
       "      <td>456</td>\n",
       "      <td>0.8185</td>\n",
       "      <td>0.305892</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                condition  site gene  gene_site  site_total escape  \\\n",
       "108   subject A (day 120)   126    S        456             3.7920   \n",
       "133   subject A (day 120)   154    S        484             3.0420   \n",
       "282   subject B (day 113)   126    S        456             3.5100   \n",
       "307   subject B (day 113)   154    S        484             3.0190   \n",
       "456   subject C (day 104)   126    S        456             0.6150   \n",
       "481   subject C (day 104)   154    S        484             0.7445   \n",
       "630    subject D (day 76)   126    S        456             0.1029   \n",
       "804   subject E (day 104)   126    S        456             3.7300   \n",
       "818   subject E (day 104)   142    S        472             3.6470   \n",
       "829   subject E (day 104)   154    S        484             8.5360   \n",
       "972   subject F (day 115)   119    S        449             3.6630   \n",
       "978   subject F (day 115)   126    S        456             5.6520   \n",
       "1003  subject F (day 115)   154    S        484             4.9770   \n",
       "1318  subject H (day 152)   117    S        447             2.2260   \n",
       "1320  subject H (day 152)   119    S        449             2.8720   \n",
       "1323  subject H (day 152)   122    S        452             2.5470   \n",
       "1351  subject H (day 152)   154    S        484             2.9980   \n",
       "1500  subject I (day 102)   126    S        456             4.7590   \n",
       "1525  subject I (day 102)   154    S        484             7.3300   \n",
       "1527  subject I (day 102)   156    S        486             5.0250   \n",
       "1530  subject I (day 102)   160    S        490             4.9600   \n",
       "1848  subject K (day 103)   126    S        456             0.8185   \n",
       "\n",
       "      upper_threshold  \n",
       "108          2.785092  \n",
       "133          2.785092  \n",
       "282          2.134808  \n",
       "307          2.134808  \n",
       "456          0.305421  \n",
       "481          0.305421  \n",
       "630          0.047514  \n",
       "804          3.191500  \n",
       "818          3.191500  \n",
       "829          3.191500  \n",
       "972          3.365717  \n",
       "978          3.365717  \n",
       "1003         3.365717  \n",
       "1318         2.224005  \n",
       "1320         2.224005  \n",
       "1323         2.224005  \n",
       "1351         2.224005  \n",
       "1500         4.384038  \n",
       "1525         4.384038  \n",
       "1527         4.384038  \n",
       "1530         4.384038  \n",
       "1848         0.305892  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "site_specific_df[site_specific_df[escape_column] > site_specific_df[\"upper_threshold\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "antigenic_epitope_sites = site_specific_df.loc[\n",
    "    site_specific_df[escape_column] > site_specific_df[\"upper_threshold\"],\n",
    "    [\"gene\", \"gene_site\"]\n",
    "].drop_duplicates().sort_values([\"gene\", \"gene_site\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1318</th>\n",
       "      <td>S</td>\n",
       "      <td>447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>S</td>\n",
       "      <td>449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1323</th>\n",
       "      <td>S</td>\n",
       "      <td>452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>S</td>\n",
       "      <td>456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>S</td>\n",
       "      <td>472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>S</td>\n",
       "      <td>484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1527</th>\n",
       "      <td>S</td>\n",
       "      <td>486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1530</th>\n",
       "      <td>S</td>\n",
       "      <td>490</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene  gene_site\n",
       "1318    S        447\n",
       "972     S        449\n",
       "1323    S        452\n",
       "108     S        456\n",
       "818     S        472\n",
       "133     S        484\n",
       "1527    S        486\n",
       "1530    S        490"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "antigenic_epitope_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8, 2)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "antigenic_epitope_sites.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "antigenic_escape_map = {\n",
    "    \"name\": \"antigenic_escape_epitope_sites\",\n",
    "    \"description\": f\"Sites with greater than 4 stddev site-level escape per serum across {len(valid_conditions)} human sera from Greaney et al. 2021\",\n",
    "    \"default\": 0,\n",
    "    \"map\": {}\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "for record in antigenic_epitope_sites.to_dict(\"records\"):\n",
    "    if record[\"gene\"] not in antigenic_escape_map[\"map\"]:\n",
    "        antigenic_escape_map[\"map\"][record[\"gene\"]] = {}\n",
    "        \n",
    "    antigenic_escape_map[\"map\"][record[\"gene\"]][str(record[\"gene_site\"])] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'antigenic_escape_epitope_sites',\n",
       " 'description': 'Sites with greater than 4 stddev site-level escape per serum across 11 human sera from Greaney et al. 2021',\n",
       " 'default': 0,\n",
       " 'map': {'S': {'447': 1,\n",
       "   '449': 1,\n",
       "   '452': 1,\n",
       "   '456': 1,\n",
       "   '472': 1,\n",
       "   '484': 1,\n",
       "   '486': 1,\n",
       "   '490': 1}}}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "antigenic_escape_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(antigenic_escape_distance_map, \"w\") as oh:\n",
    "    json.dump(antigenic_escape_map, oh, indent=4, sort_keys=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/jlhudd/projects/nextstrain/ncov\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identify weighted antigenic escape epitope sites\n",
    "\n",
    "Instead of using an arbitrary threshold (e.g., 4 std dev from the mean), calculate the average proportion of the total escape value at each position in S across all sera. The resulting distance map weights each position in S by its relative contribution to antigenic escape."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_escape = site_specific_df.groupby(\"condition\").aggregate({escape_column: \"sum\"}).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_escape = total_escape.rename(columns={escape_column: \"total_escape\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>total_escape</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>92.674160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>subject B (day 113)</td>\n",
       "      <td>60.538330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>subject C (day 104)</td>\n",
       "      <td>2.690213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>subject D (day 76)</td>\n",
       "      <td>1.634826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>subject E (day 104)</td>\n",
       "      <td>22.716179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>subject F (day 115)</td>\n",
       "      <td>62.018851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>subject G (day 94)</td>\n",
       "      <td>95.918280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>subject H (day 152)</td>\n",
       "      <td>42.703905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>subject I (day 102)</td>\n",
       "      <td>66.123663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>subject J (day 121)</td>\n",
       "      <td>86.877970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>subject K (day 103)</td>\n",
       "      <td>6.365402</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              condition  total_escape\n",
       "0   subject A (day 120)     92.674160\n",
       "1   subject B (day 113)     60.538330\n",
       "2   subject C (day 104)      2.690213\n",
       "3    subject D (day 76)      1.634826\n",
       "4   subject E (day 104)     22.716179\n",
       "5   subject F (day 115)     62.018851\n",
       "6    subject G (day 94)     95.918280\n",
       "7   subject H (day 152)     42.703905\n",
       "8   subject I (day 102)     66.123663\n",
       "9   subject J (day 121)     86.877970\n",
       "10  subject K (day 103)      6.365402"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_escape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "site_specific_df = site_specific_df.merge(\n",
    "    total_escape,\n",
    "    on=\"condition\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>site</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_site</th>\n",
       "      <th>site_total escape</th>\n",
       "      <th>upper_threshold</th>\n",
       "      <th>total_escape</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>1</td>\n",
       "      <td>S</td>\n",
       "      <td>331</td>\n",
       "      <td>0.1205</td>\n",
       "      <td>2.785092</td>\n",
       "      <td>92.67416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>2</td>\n",
       "      <td>S</td>\n",
       "      <td>332</td>\n",
       "      <td>0.6332</td>\n",
       "      <td>2.785092</td>\n",
       "      <td>92.67416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>3</td>\n",
       "      <td>S</td>\n",
       "      <td>333</td>\n",
       "      <td>0.1667</td>\n",
       "      <td>2.785092</td>\n",
       "      <td>92.67416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>4</td>\n",
       "      <td>S</td>\n",
       "      <td>334</td>\n",
       "      <td>0.5097</td>\n",
       "      <td>2.785092</td>\n",
       "      <td>92.67416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>subject A (day 120)</td>\n",
       "      <td>5</td>\n",
       "      <td>S</td>\n",
       "      <td>335</td>\n",
       "      <td>0.6866</td>\n",
       "      <td>2.785092</td>\n",
       "      <td>92.67416</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             condition  site gene  gene_site  site_total escape  \\\n",
       "0  subject A (day 120)     1    S        331             0.1205   \n",
       "1  subject A (day 120)     2    S        332             0.6332   \n",
       "2  subject A (day 120)     3    S        333             0.1667   \n",
       "3  subject A (day 120)     4    S        334             0.5097   \n",
       "4  subject A (day 120)     5    S        335             0.6866   \n",
       "\n",
       "   upper_threshold  total_escape  \n",
       "0         2.785092      92.67416  \n",
       "1         2.785092      92.67416  \n",
       "2         2.785092      92.67416  \n",
       "3         2.785092      92.67416  \n",
       "4         2.785092      92.67416  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "site_specific_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sum total escape values per site and condition and calculate the proportion of these totals per site. This approach averages the antigenic escape effect per site across all conditions (sera)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_site_specific_escape = site_specific_df.groupby([\"gene\", \"gene_site\"]).aggregate({\n",
    "    escape_column: \"sum\",\n",
    "    \"total_escape\": \"sum\"\n",
    "}).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_site_specific_escape[\"proportion_of_escape\"] = (\n",
    "    total_site_specific_escape[escape_column] / total_site_specific_escape[\"total_escape\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_site</th>\n",
       "      <th>site_total escape</th>\n",
       "      <th>total_escape</th>\n",
       "      <th>proportion_of_escape</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S</td>\n",
       "      <td>331</td>\n",
       "      <td>1.478387</td>\n",
       "      <td>540.261779</td>\n",
       "      <td>0.002736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S</td>\n",
       "      <td>332</td>\n",
       "      <td>3.153901</td>\n",
       "      <td>540.261779</td>\n",
       "      <td>0.005838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S</td>\n",
       "      <td>333</td>\n",
       "      <td>1.640904</td>\n",
       "      <td>540.261779</td>\n",
       "      <td>0.003037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S</td>\n",
       "      <td>334</td>\n",
       "      <td>2.693905</td>\n",
       "      <td>540.261779</td>\n",
       "      <td>0.004986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S</td>\n",
       "      <td>335</td>\n",
       "      <td>3.239830</td>\n",
       "      <td>540.261779</td>\n",
       "      <td>0.005997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>S</td>\n",
       "      <td>527</td>\n",
       "      <td>2.846798</td>\n",
       "      <td>540.261779</td>\n",
       "      <td>0.005269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>S</td>\n",
       "      <td>528</td>\n",
       "      <td>3.524410</td>\n",
       "      <td>540.261779</td>\n",
       "      <td>0.006524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>S</td>\n",
       "      <td>529</td>\n",
       "      <td>3.942250</td>\n",
       "      <td>540.261779</td>\n",
       "      <td>0.007297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>S</td>\n",
       "      <td>530</td>\n",
       "      <td>3.148190</td>\n",
       "      <td>540.261779</td>\n",
       "      <td>0.005827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>S</td>\n",
       "      <td>531</td>\n",
       "      <td>3.027686</td>\n",
       "      <td>540.261779</td>\n",
       "      <td>0.005604</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>174 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene  gene_site  site_total escape  total_escape  proportion_of_escape\n",
       "0      S        331           1.478387    540.261779              0.002736\n",
       "1      S        332           3.153901    540.261779              0.005838\n",
       "2      S        333           1.640904    540.261779              0.003037\n",
       "3      S        334           2.693905    540.261779              0.004986\n",
       "4      S        335           3.239830    540.261779              0.005997\n",
       "..   ...        ...                ...           ...                   ...\n",
       "169    S        527           2.846798    540.261779              0.005269\n",
       "170    S        528           3.524410    540.261779              0.006524\n",
       "171    S        529           3.942250    540.261779              0.007297\n",
       "172    S        530           3.148190    540.261779              0.005827\n",
       "173    S        531           3.027686    540.261779              0.005604\n",
       "\n",
       "[174 rows x 5 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_site_specific_escape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_site</th>\n",
       "      <th>site_total escape</th>\n",
       "      <th>total_escape</th>\n",
       "      <th>proportion_of_escape</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>S</td>\n",
       "      <td>484</td>\n",
       "      <td>33.45514</td>\n",
       "      <td>540.261779</td>\n",
       "      <td>0.061924</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene  gene_site  site_total escape  total_escape  proportion_of_escape\n",
       "133    S        484           33.45514    540.261779              0.061924"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_site_specific_escape.query(\"gene == 'S' & gene_site == 484\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "weighted_antigenic_escape_map = {\n",
    "    \"name\": \"antigenic_escape_weighted_epitope_sites\",\n",
    "    \"description\": \"Average proportion of site-level total escape per site across four human sera from Greaney et al. 2021\",\n",
    "    \"default\": 0.0,\n",
    "    \"map\": {}\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "for record in total_site_specific_escape.to_dict(\"records\"):\n",
    "    if record[\"gene\"] not in weighted_antigenic_escape_map[\"map\"]:\n",
    "        weighted_antigenic_escape_map[\"map\"][record[\"gene\"]] = {}\n",
    "        \n",
    "    weighted_antigenic_escape_map[\"map\"][record[\"gene\"]][str(record[\"gene_site\"])] = record[\"proportion_of_escape\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': 'antigenic_escape_weighted_epitope_sites',\n",
       " 'description': 'Average proportion of site-level total escape per site across four human sera from Greaney et al. 2021',\n",
       " 'default': 0.0,\n",
       " 'map': {'S': {'331': 0.0027364271507114802,\n",
       "   '332': 0.005837727419854265,\n",
       "   '333': 0.003037238732017442,\n",
       "   '334': 0.004986295728680927,\n",
       "   '335': 0.0059967780937532425,\n",
       "   '336': 0.0011751290298753962,\n",
       "   '337': 0.004510424567535593,\n",
       "   '338': 0.0015981696908446933,\n",
       "   '339': 0.005228752265752545,\n",
       "   '340': 0.0026803728429881666,\n",
       "   '341': 0.0008037307413537133,\n",
       "   '342': 0.0007009435330426895,\n",
       "   '344': 0.0015493600192470253,\n",
       "   '345': 0.0014853558617128663,\n",
       "   '346': 0.0065186454755736655,\n",
       "   '347': 0.000519251797939699,\n",
       "   '348': 0.0033662496059586144,\n",
       "   '349': 0.0007261368754816679,\n",
       "   '351': 0.0011045238131141325,\n",
       "   '352': 0.0030244301265014823,\n",
       "   '353': 0.00029253244667990197,\n",
       "   '354': 0.0030305234688943355,\n",
       "   '356': 0.004199971363956128,\n",
       "   '357': 0.005834634474793981,\n",
       "   '358': 0.0025016539259948853,\n",
       "   '359': 0.005084048340604175,\n",
       "   '360': 0.002196649932623366,\n",
       "   '361': 0.002341520443681625,\n",
       "   '362': 0.006745489210979512,\n",
       "   '363': 0.004072070404251962,\n",
       "   '365': 0.005305500245393262,\n",
       "   '366': 0.006027096729353158,\n",
       "   '367': 0.007408186470066092,\n",
       "   '368': 0.0010608375837228485,\n",
       "   '369': 0.009108658419128574,\n",
       "   '370': 0.004383737834019066,\n",
       "   '371': 0.004886234976428431,\n",
       "   '372': 0.005183209528943267,\n",
       "   '373': 0.005976325045927902,\n",
       "   '374': 0.00398462390728722,\n",
       "   '375': 0.002444252123356019,\n",
       "   '376': 0.003538906646786467,\n",
       "   '377': 0.0015250495821304616,\n",
       "   '378': 0.004714934685288902,\n",
       "   '380': 0.0005304110178522961,\n",
       "   '381': 0.0017751283500568077,\n",
       "   '382': 0.0019932540895117637,\n",
       "   '383': 0.008476279795641913,\n",
       "   '384': 0.0076845895136641125,\n",
       "   '385': 0.00844862653460023,\n",
       "   '386': 0.007586951660923232,\n",
       "   '387': 0.0005396844852649421,\n",
       "   '388': 0.003890267796971167,\n",
       "   '389': 0.003684650808394369,\n",
       "   '390': 0.005829229687495339,\n",
       "   '391': 0.0054953545049853895,\n",
       "   '392': 0.0042747869470421244,\n",
       "   '393': 0.005165854979041874,\n",
       "   '394': 0.005753320190267732,\n",
       "   '395': 0.0015734575965898404,\n",
       "   '396': 0.004284030614086447,\n",
       "   '397': 0.001343830025534281,\n",
       "   '399': 0.0002730905753275916,\n",
       "   '401': 0.0004539266141401154,\n",
       "   '402': 0.0010310668306710133,\n",
       "   '403': 0.0018837368104411979,\n",
       "   '404': 0.000500035372852106,\n",
       "   '405': 0.0012881607163582676,\n",
       "   '406': 0.003754798284094348,\n",
       "   '407': 0.0006468118118149579,\n",
       "   '408': 0.005103764708516892,\n",
       "   '409': 0.0004515342553786446,\n",
       "   '410': 0.0002623554091033915,\n",
       "   '411': 0.0009502300924197823,\n",
       "   '413': 0.004271392296389485,\n",
       "   '414': 0.0033306594517879674,\n",
       "   '415': 0.003641717917506697,\n",
       "   '417': 0.013036346964324621,\n",
       "   '418': 0.00024254427231749231,\n",
       "   '419': 0.00048489086268858233,\n",
       "   '420': 0.002635490896954785,\n",
       "   '421': 0.0011066487089425026,\n",
       "   '424': 0.00043046132287305916,\n",
       "   '425': 0.0007342699697933915,\n",
       "   '427': 0.004983987588351679,\n",
       "   '428': 0.0029127823987388835,\n",
       "   '429': 0.000250985180371601,\n",
       "   '430': 0.004278847941334324,\n",
       "   '433': 0.00022324159274766746,\n",
       "   '434': 0.0014802546309611341,\n",
       "   '435': 0.00046967731192018206,\n",
       "   '437': 0.0006413241017522819,\n",
       "   '438': 0.000516706735427496,\n",
       "   '439': 0.004956591609993048,\n",
       "   '440': 0.00557844385492924,\n",
       "   '441': 0.0024230549917998384,\n",
       "   '443': 0.006783692912980872,\n",
       "   '444': 0.011932696801760132,\n",
       "   '445': 0.011183282321803219,\n",
       "   '446': 0.01674417912755741,\n",
       "   '447': 0.02634881007429134,\n",
       "   '448': 0.018340497863847775,\n",
       "   '449': 0.03043052580272591,\n",
       "   '450': 0.016391720361321995,\n",
       "   '451': 0.0019886544674442555,\n",
       "   '452': 0.020266654850765096,\n",
       "   '453': 0.005101885989644248,\n",
       "   '455': 0.01535729589908943,\n",
       "   '456': 0.051125956126955995,\n",
       "   '458': 0.004551987752053062,\n",
       "   '459': 0.005516120364130412,\n",
       "   '460': 0.007323968037844101,\n",
       "   '461': 0.000878666636485742,\n",
       "   '462': 0.004705159794287488,\n",
       "   '463': 0.0002609003737282331,\n",
       "   '464': 0.002966397148359591,\n",
       "   '465': 0.0010053496680931596,\n",
       "   '466': 0.0009203112630036008,\n",
       "   '468': 0.005579017650841082,\n",
       "   '469': 0.0027374914495802193,\n",
       "   '470': 0.0064934947050894355,\n",
       "   '471': 0.005131156984966414,\n",
       "   '472': 0.02185499782388085,\n",
       "   '473': 0.016601507550509695,\n",
       "   '474': 0.0031918174997131593,\n",
       "   '475': 0.0089894754553753,\n",
       "   '476': 0.003652396074330624,\n",
       "   '477': 0.004557459173715659,\n",
       "   '478': 0.0051832521008980175,\n",
       "   '479': 0.004933034511380097,\n",
       "   '481': 0.005241736711951165,\n",
       "   '482': 0.005041957634038724,\n",
       "   '483': 0.010863159731631936,\n",
       "   '484': 0.06192394374872998,\n",
       "   '485': 0.015918782963145275,\n",
       "   '486': 0.02485165622824918,\n",
       "   '487': 0.008105676121910403,\n",
       "   '489': 0.007163608739075213,\n",
       "   '490': 0.025692063634096932,\n",
       "   '492': 0.0013016763865139815,\n",
       "   '493': 0.00934335945661755,\n",
       "   '494': 0.013567089673233054,\n",
       "   '496': 0.01844659457890194,\n",
       "   '498': 0.008317334626152532,\n",
       "   '499': 0.009427081833018982,\n",
       "   '500': 0.006127714618926508,\n",
       "   '501': 0.009443009296958989,\n",
       "   '503': 0.009975997213741821,\n",
       "   '504': 0.007076121521110277,\n",
       "   '505': 0.0033997703929375205,\n",
       "   '506': 0.0015954839557863613,\n",
       "   '507': 0.00030536104250504867,\n",
       "   '508': 0.0027582924768617748,\n",
       "   '510': 0.00019822612704136015,\n",
       "   '512': 0.00022566042015926524,\n",
       "   '513': 0.00040200604322298585,\n",
       "   '514': 0.004630851372749376,\n",
       "   '515': 0.000286219210885995,\n",
       "   '516': 0.002244476747352686,\n",
       "   '517': 0.003983219032780485,\n",
       "   '518': 0.005881744970110773,\n",
       "   '519': 0.004559245344860586,\n",
       "   '520': 0.005649052218313246,\n",
       "   '521': 0.002315536743647948,\n",
       "   '522': 0.004826649047415457,\n",
       "   '523': 0.00044379226776425075,\n",
       "   '524': 0.0006798222906232359,\n",
       "   '525': 0.0017546623455495867,\n",
       "   '526': 0.0034884089046352505,\n",
       "   '527': 0.005269293723356022,\n",
       "   '528': 0.006523522740824323,\n",
       "   '529': 0.007296925591805347,\n",
       "   '530': 0.00582715661839449,\n",
       "   '531': 0.005604109190779571}}}"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "weighted_antigenic_escape_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(weighted_antigenic_escape_distance_map, \"w\") as oh:\n",
    "    json.dump(weighted_antigenic_escape_map, oh, indent=4, sort_keys=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate antigenic escape effects for specific mutations\n",
    "\n",
    "Follow the approach from Lee et al. 2018 where mutational effects were estimated by the log ratio of DMS preference values of the derived and ancestral amino acids at a given site.\n",
    "\n",
    "Build a distance map for each mutation from one amino acid to another using this equation. For any mutations that are not defined (e.g., mutations to wildtype at a site), the preference is zero."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "escape_by_mutation = df.query(\"condition == 'subject H (day 152)'\").loc[\n",
    "    :, [\"condition\", \"label_site\", \"wildtype\", \"mutation\", \"mut_escape color ACE2 bind\"]\n",
    "].dropna().rename(\n",
    "    columns={\"mut_escape color ACE2 bind\": \"mut_escape\"}\n",
    ").groupby([\n",
    "    \"label_site\",\n",
    "    \"wildtype\",\n",
    "    \"mutation\"\n",
    "])[\"mut_escape\"].mean().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label_site</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mut_escape</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>0.002020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "      <td>0.005616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>E</td>\n",
       "      <td>0.002535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>F</td>\n",
       "      <td>0.003032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>331</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>0.003113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>531</td>\n",
       "      <td>T</td>\n",
       "      <td>R</td>\n",
       "      <td>0.003578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1942</th>\n",
       "      <td>531</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>0.008192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1943</th>\n",
       "      <td>531</td>\n",
       "      <td>T</td>\n",
       "      <td>V</td>\n",
       "      <td>0.002950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1944</th>\n",
       "      <td>531</td>\n",
       "      <td>T</td>\n",
       "      <td>W</td>\n",
       "      <td>0.002530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1945</th>\n",
       "      <td>531</td>\n",
       "      <td>T</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.004839</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1946 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      label_site wildtype mutation  mut_escape\n",
       "0            331        N        A    0.002020\n",
       "1            331        N        D    0.005616\n",
       "2            331        N        E    0.002535\n",
       "3            331        N        F    0.003032\n",
       "4            331        N        G    0.003113\n",
       "...          ...      ...      ...         ...\n",
       "1941         531        T        R    0.003578\n",
       "1942         531        T        S    0.008192\n",
       "1943         531        T        V    0.002950\n",
       "1944         531        T        W    0.002530\n",
       "1945         531        T        Y    0.004839\n",
       "\n",
       "[1946 rows x 4 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "escape_by_mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = \"S\"\n",
    "mutation_antigenic_escape_map = {\n",
    "    \"name\": \"antigenic_escape_mutation_effect\",\n",
    "    \"default\": 0.0,\n",
    "    \"map\": {\n",
    "        gene: {}\n",
    "    }\n",
    "}\n",
    "\n",
    "for index, row in escape_by_mutation.iterrows():\n",
    "    position = str(row[\"label_site\"])\n",
    "            \n",
    "    if position not in mutation_antigenic_escape_map[\"map\"][gene]:\n",
    "        mutation_antigenic_escape_map[\"map\"][gene][position] = []\n",
    "    \n",
    "    mutation_antigenic_escape_map[\"map\"][gene][position].append({\n",
    "        \"from\": row[\"wildtype\"],\n",
    "        \"to\": row[\"mutation\"],\n",
    "        \"weight\": row[\"mut_escape\"]\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'from': 'E', 'to': 'A', 'weight': 0.2072},\n",
       " {'from': 'E', 'to': 'C', 'weight': 0.1541},\n",
       " {'from': 'E', 'to': 'D', 'weight': 0.01901},\n",
       " {'from': 'E', 'to': 'F', 'weight': 0.2066},\n",
       " {'from': 'E', 'to': 'G', 'weight': 0.05396},\n",
       " {'from': 'E', 'to': 'H', 'weight': 0.2303},\n",
       " {'from': 'E', 'to': 'I', 'weight': 0.2216},\n",
       " {'from': 'E', 'to': 'K', 'weight': 0.14300000000000002},\n",
       " {'from': 'E', 'to': 'L', 'weight': 0.1879},\n",
       " {'from': 'E', 'to': 'M', 'weight': 0.1632},\n",
       " {'from': 'E', 'to': 'N', 'weight': 0.04948},\n",
       " {'from': 'E', 'to': 'P', 'weight': 0.2348},\n",
       " {'from': 'E', 'to': 'Q', 'weight': 0.1674},\n",
       " {'from': 'E', 'to': 'R', 'weight': 0.1387},\n",
       " {'from': 'E', 'to': 'S', 'weight': 0.1491},\n",
       " {'from': 'E', 'to': 'T', 'weight': 0.1291},\n",
       " {'from': 'E', 'to': 'V', 'weight': 0.21},\n",
       " {'from': 'E', 'to': 'W', 'weight': 0.1513},\n",
       " {'from': 'E', 'to': 'Y', 'weight': 0.1815}]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutation_antigenic_escape_map[\"map\"][\"S\"][\"484\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(mutation_antigenic_escape_distance_map, \"w\") as oh:\n",
    "    json.dump(mutation_antigenic_escape_map, oh, indent=4, sort_keys=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
