Compiled Nextstrain SARS-CoV-2 resources are available at [nextstrain.org/sars-cov-2](https://nextstrain.org/sars-cov-2/). Follow [@nextstrain](https://twitter.com/nextstrain) for updates.

This phylogeny shows evolutionary relationships of SARS-CoV-2 viruses relative to clade 21L (Pango lineage BA.2) reference virus. Although the genetic relationships among sampled viruses are generally quite clear, there is considerable uncertainty surrounding estimates of specific transmission dates and in reconstruction of geographic spread. Please be aware that specific inferred geographic transmission patterns and temporal estimates are only a hypothesis.

There are millions of complete SARS-CoV-2 genomes available on open databases and this number increases every day. This visualization can only handle ~4000 genomes in a single view for performance and legibility reasons. Because of this we subsample available genome data for our analysis views. We provision multiple views to focus subsampling with different reference viruses, different geographic regions and different time periods. These views are available through the "Dataset" dropdown on the left or by clicking on the following links:

**Using ancestral Wuhan viruses as phylogeny root**
&nbsp;            | past 1 month                                                               | past 2 months                                                              | past 6 months                                                              | all time
----------------- | -------------------------------------------------------------------------- | -------------------------------------------------------------------------- | -------------------------------------------------------------------------- | --------------------------------------------------------------------------------------
**global**        | [global/1m](/ncov/open/global/1m)                                        | [global/2m](/ncov/open/global/2m)                                        | [global/6m](/ncov/open/global/6m)                                        | [global/all-time](/ncov/open/global/all-time)                                        |
**Africa**        | [africa/1m](/ncov/open/africa/1m?f_region=Africa)                        | [africa/2m](/ncov/open/africa/2m?f_region=Africa)                        | [africa/6m](/ncov/open/africa/6m?f_region=Africa)                        | [africa/all-time](/ncov/open/africa/all-time?f_region=Africa)                        |
**Asia**          | [asia/1m](/ncov/open/asia/1m?f_region=Asia)                              | [asia/2m](/ncov/open/asia/2m?f_region=Asia)                              | [asia/6m](/ncov/open/asia/6m?f_region=Asia)                              | [asia/all-time](/ncov/open/asia/all-time?f_region=Asia)                              |
**Europe**        | [europe/1m](/ncov/open/europe/1m?f_region=Europe)                        | [europe/2m](/ncov/open/europe/2m?f_region=Europe)                        | [europe/6m](/ncov/open/europe/6m?f_region=Europe)                        | [europe/all-time](/ncov/open/europe/all-time?f_region=Europe)                        |
**North America** | [north-america/1m](/ncov/open/north-america/1m?f_region=North%20America) | [north-america/2m](/ncov/open/north-america/2m?f_region=North%20America) | [north-america/6m](/ncov/open/north-america/6m?f_region=North%20America) | [north-america/all-time](/ncov/open/north-america/all-time?f_region=North%20America) |
**Oceania**       | [oceania/1m](/ncov/open/oceania/1m?f_region=Oceania)                     | [oceania/2m](/ncov/open/oceania/2m?f_region=Oceania)                     | [oceania/6m](/ncov/open/oceania/6m?f_region=Oceania)                     | [oceania/all-time](/ncov/open/oceania/all-time?f_region=Oceania)                     |
**South America** | [south-america/1m](/ncov/open/south-america/1m?f_region=South%20America) | [south-america/2m](/ncov/open/south-america/2m?f_region=South%20America) | [south-america/6m](/ncov/open/south-america/6m?f_region=South%20America) | [south-america/all-time](/ncov/open/south-america/all-time?f_region=South%20America) |

**Using clade 21L (lineage BA.2) as phylogeny root**
&nbsp;            | past 1 month                                                                       | past 2 months                                                                      | past 6 months                                                                      | all time
----------------- | ---------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------
**global**        | [21L/global/1m](/ncov/open/21L/global/1m)                                        | [21L/global/2m](/ncov/open/21L/global/2m)                                        | [21L/global/6m](/ncov/open/21L/global/6m)                                        | [21L/global/all-time](/ncov/open/21L/global/all-time)                                        |
**Africa**        | [21L/africa/1m](/ncov/open/21L/africa/1m?f_region=Africa)                        | [21L/africa/2m](/ncov/open/21L/africa/2m?f_region=Africa)                        | [21L/africa/6m](/ncov/open/21L/africa/6m?f_region=Africa)                        | [21L/africa/all-time](/ncov/open/21L/africa/all-time?f_region=Africa)                        |
**Asia**          | [21L/asia/1m](/ncov/open/21L/asia/1m?f_region=Asia)                              | [21L/asia/2m](/ncov/open/21L/asia/2m?f_region=Asia)                              | [21L/asia/6m](/ncov/open/21L/asia/6m?f_region=Asia)                              | [21L/asia/all-time](/ncov/open/21L/asia/all-time?f_region=Asia)                              |
**Europe**        | [21L/europe/1m](/ncov/open/21L/europe/1m?f_region=Europe)                        | [21L/europe/2m](/ncov/open/21L/europe/2m?f_region=Europe)                        | [21L/europe/6m](/ncov/open/21L/europe/6m?f_region=Europe)                        | [21L/europe/all-time](/ncov/open/21L/europe/all-time?f_region=Europe)                        |
**North America** | [21L/north-america/1m](/ncov/open/21L/north-america/1m?f_region=North%20America) | [21L/north-america/2m](/ncov/open/21L/north-america/2m?f_region=North%20America) | [21L/north-america/6m](/ncov/open/21L/north-america/6m?f_region=North%20America) | [21L/north-america/all-time](/ncov/open/21L/north-america/all-time?f_region=North%20America) |
**Oceania**       | [21L/oceania/1m](/ncov/open/21L/oceania/1m?f_region=Oceania)                     | [21L/oceania/2m](/ncov/open/21L/oceania/2m?f_region=Oceania)                     | [21L/oceania/6m](/ncov/open/21L/oceania/6m?f_region=Oceania)                     | [21L/oceania/all-time](/ncov/open/21L/oceania/all-time?f_region=Oceania)                     |
**South America** | [21L/south-america/1m](/ncov/open/21L/south-america/1m?f_region=South%20America) | [21L/south-america/2m](/ncov/open/21L/south-america/2m?f_region=South%20America) | [21L/south-america/6m](/ncov/open/21L/south-america/6m?f_region=South%20America) | [21L/south-america/all-time](/ncov/open/21L/south-america/all-time?f_region=South%20America) |

Site numbering and genome structure uses [Wuhan-Hu-1/2019](https://www.ncbi.nlm.nih.gov/nuccore/MN908947) as reference. The phylogeny is rooted relative to clade 21L (Pango lineage BA.2) reference virus. Temporal resolution assumes a nucleotide substitution rate of 8 &times; 10^-4 subs per site per year. Mutational fitness is calculated using results from [Obermeyer et al](https://doi.org/10.1126/science.abm1208). Immune escape vs BA.2 is estimated using the [RBD antibody escape calculator](https://jbloomlab.github.io/SARS2-RBD-escape-calc/) maintained by Jesse Bloom. Full details on bioinformatic processing can be found [here](https://github.com/nextstrain/ncov).

We gratefully acknowledge the authors, originating and submitting laboratories of the genetic sequences and metadata for sharing their work in open databases. Please note that although data generators have generously shared data in an open fashion, that does not mean there should be free license to publish on this data. Data generators should be cited where possible and collaborations should be sought in some circumstances. Please try to avoid scooping someone else's work. Reach out if uncertain. An attribution table is available by clicking on "Download Data" at the bottom of the page and then clicking on "Strain Metadata" in the resulting dialog box.

To maximize the utility and visibility of these generously shared data, [we provide preprocessed files that can serve as a starting point for additional analyses](https://docs.nextstrain.org/projects/ncov/en/latest/reference/remote_inputs.html).

#### Reusing code or images

All source code for Auspice, the visualization tool, is freely available under the terms of the [GNU Affero General Public License 3.0](https://github.com/nextstrain/auspice/blob/HEAD/LICENSE.txt).

Screenshots may be used under a [CC-BY-4.0 license](https://creativecommons.org/licenses/by/4.0/) and attribution to nextstrain.org must be provided. A high-quality download option is available by clicking the **DOWNLOAD DATA** button at the bottom of the page and selecting **SCREENSHOT (SVG)**.
