---
title: Genomic analysis of COVID-19 spread. Situation report 2020-05-01.
authors:
  - Sidney Bell  
  - Nicola MÃ¼ller
  - Cassia Wagner
  - Emma Hodcroft
  - James Hadfield
  - Sidney M. Bell
  - Richard Neher
  - Trevor Bedford
authorLinks:
  - https://twitter.com/sidneymbell
  - https://bedford.io/team/nicola-mueller/
  - https://bedford.io/team/cassia-wagner/
  - https://neherlab.org/emma-hodcroft.html/
  - https://bedford.io/team/james-hadfield/
  - https://neherlab.org/richard-neher.html/
  - https://bedford.io/team/trevor-bedford/
affiliations: "Fred Hutch, Seattle, USA; Biozentrum, Basel, Switzerland; CZI, CA, USA"
translators:
translatorLinks:
license: "CC-BY"  
licenseLink: "https://creativecommons.org/licenses/by/4.0/"
dataset: "https://nextstrain.org/ncov/europe/2020-04-30?d=map&legend=closed"

abstract: "This weekly report uses publicly shared genomic data to track the spread of COVID-19. This week, we focus on outbreak dynamics in Central and Eastern Europe."
---
<!-- Translators: Only text after : in the above ^ needs to be translated -->
<!-- Comment tags like these do not need to be translated, they are only to help you! -->
<!-- Ensure that links always end in a 'letter' (. counts) If some kind of text doesn't follow them, it breaks the slide. -->
<!-- numbers can be tagged ilke this: 161</tag> - this is just for us to help find them to update! Just leave in the </tag> bit. -->

<!-- This is left-side text -->
# [Table of Contents](https://nextstrain.org/ncov/2020-XX-XX?d=tree,map&p=grid)

* [Background resources](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=2).     
* [About this data](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=3).
* [Updates for Russia](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=X).
* [Updates for Turkey](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=X).
* [Updates for the Czech Republic](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=X).
* [Updates for Poland](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=X).
* [Data from Hungary, Slovakia, Slovenia and Latvia](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=X).
* [What you can do](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=X).
* [FAQ & common misconceptions](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=X).
* [Scientific credit](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=X).

<!-- This is right-side text -->
```auspiceMainDisplayMarkdown
# Executive summary

We analyzed 4,442 publicly shared COVID-19 genomes. By comparing these viral genomes to each other, we can characterize how COVID-19 is moving around the world and spreading locally. In this week's updates, we report:

* Many introductions to Russia  
* Heterogeneous outbreak in Moscow  
* Majority of St. Petersburg outbreak tied to a single introduction, most likely from Western Europe  
* Extensive mixing of cases across Turkey following repeated travel-related introductions  
* At least 5 local transmission chains in the Czech Republic, mostly likely due to repeated introductions from Western Europe
* At least 11 separate introductions to Poland
* Early samples from Hungary, Latvia, Poland, Slovakia and Slovenia  
```


<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text 2-->
# [COVID-19 Resources](https://nextstrain.org/ncov/XX/2020-XX-XX?d=tree&p=full&legend=closed)
We've prepared some resources that will make interpreting the data we present in this narrative easier.
#### Nextstrain Resources  
* [START HERE: How to read a phylogeny](https://nextstrain.org/narratives/trees-background/).  
* [Background on coronaviruses](https://nextstrain.org/help/coronavirus/human-CoV).
* [Common misconceptions](https://nextstrain.org/narratives/ncov/sit-rep/2020-03-13?n=11).

#### External Resources  
* [Intuition: how SARS-CoV-2 mutates and spreads](https://www.nytimes.com/interactive/2020/04/30/science/coronavirus-mutations.html).
* [Ask a Scientist & FAQs](https://covid19.fas.org/l/en).
* [WHO Situation Reports](https://www.who.int/emergencies/diseases/novel-coronavirus-2019/situation-reports).
* [CDC Resources](https://www.cdc.gov/coronavirus/2019-ncov/index.html).
* [Making sense of uncertainty](https://www.theatlantic.com/health/archive/2020/04/pandemic-confusing-uncertainty/610819/).
<!-- There is no right-side text -->


<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [A note on sampling](https://nextstrain.org/ncov/europe/2020-04-30?c=country&r=country&d=map&p=grid&legend=closed)
We currently have sequences from samples collected across 6 continents. This is an incredible achievement -- sequencing an unknown, large RNA virus in the midst of a pandemic is difficult, and is only possible through the incredible work and timely sharing of data by scientists and physicians around the world.
<br><br>
While this data enables us to infer many useful characteristics of the outbreak and track its spread in real time, it's important to emphasize that our conclusions are limited by the available data.
<br><br>
For example, the map shows many samples from Iceland, the U.K., and the Netherlands. This does not necessarily indicate that SARS-CoV-2 is circulating more broadly in these areas; rather, we just have more data available from these countries. The size of each circle on the map indicates how much data is currently available from that area, rather than the true size of the outbreak.

<!-- There is NO right-side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Many introductions to Russia](https://nextstrain.org/ncov/europe/2020-04-29?c=division&d=tree,map&f_country=Russia&p=full&r=division)

Highlighting all of the samples from Russia, we see that they are scattered all across Russia geographically and all across the tree genomically.
<br><br>
This indicates that the epidemic in Russia was sparked by many separate introductions, rather than a single introduction followed by within-country spread.
<br><br>
While there are a few samples towards the bottom of the tree that are most closely related to Asian cases, the majority of Russian samples descend from European lineages.

<!-- There is no right side text -->

<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Heterogeneous outbreak in Moscow](https://nextstrain.org/ncov/europe/2020-04-29?c=division&d=tree&f_country=Russia&f_division=Moscow&label=clade:A2&p=full&r=division&m=div)

An example of this pattern can be seen by zooming in on samples from Moscow.
There are a few clusters of cases that are most closely related to other samples from Moscow, indicating likely local transmission.
<br><br>
However, most of these cases are most closely related to samples from elsewhere, consistent with importation.
Given the number of confirmed cases in Moscow, it's likely that at least some of these introductions led to further local spread, but we don't yet have enough data to see this.

<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Large outbreak cluster in St. Petersburg, sparked in early March](https://nextstrain.org/ncov/europe/2020-04-29?c=division&d=tree,map&f_country=Russia&f_division=Saint-Petersburg&f_region=Europe&label=clade:A2&p=full&r=division)

We can look at St. Petersburg as a counterexample.
<br><br>
As with Moscow, there are samples from St. Petersburg across the tree, indicating that there have been many introductions.
However, if you 'Explore the Data Yourself' (top right) and zoom into the main cluster in the center of this view, we can see a large clade that is made up entirely of samples from St. Petersburg.
<br><br>
This pattern is consistent with at least one of those introductions to St. Petersburg taking off to spark widespread transmission across the city, starting around the first week of March. This large outbreak cluster is most closely related to samples from Western Europe, although we cannot rule out the possibility of an unsampled, intermediate transmission from elsewhere.
<!-- There is no right side text -->

<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Five local transmission chains in the Czech Republic](https://nextstrain.org/ncov/europe/2020-04-29?d=tree,map&f_country=Czech%20Republic&m=div&p=full)

The data from the Czech Republic show a mix of these two patterns.
<br><br>
There are at least 7 instances where a sample isolated in the Czech Republic descends from a virus that was mostly likely circulating elsewhere, indicating a new introduction. These all descend from other European lineages, and were sampled between March 1 and April 2.
<br><br>
For 5 of these introductions, we also see closely related local cases, sampled over time. This is consistent with local transmission.
<br><br>
From this, we can say that there were at least 7 introductions to the Czech Republic; at least 5 of these likely led to further local transmission.

<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Extensive mixing of cases across Turkey](https://nextstrain.org/ncov/europe/2020-04-29?c=country_exposure&d=tree,map&f_country=Turkey&label=clade:A2&p=full&r=division)

In the middle of the tree, we see a string of cases sampled from many different provinces around Turkey.
Some of these cases have reported travel history to Saudi Arabia, with subsequent closely related cases in Turkey.
<br><br>
Normally, we might assume that returning travelers had passed the virus to household members or neighbors back home.
Interestingly, though, the travelers' cases were often sampled from different provinces than the most closely related cases.
<br><br>
This indicates that there has been extensive mixing of viral lineages which reached many provinces of Turkey within one week.
We may also speculate that a large gathering could explain these observations, although we would need additional epidemiological data to further investigate.
<!-- There is no right-side text -->


<!-- There is no right-side text -->
<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [At least 11 introductions to Poland](https://nextstrain.org/ncov/europe/2020-04-29?d=tree,map&f_country=Poland&p=full)

We have 15 sequences from Poland on Nextstrain. Of these 15, at least 11 of them represent separate introductions from elsewhere in the world.
As in Moscow, given the number of reported cases, we might expect that several of these introductions have seeded local transmission chains in Poland, but we can't yet see this in the genomic data.
<!-- There is no right-side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Early samples from Hungary, Latvia, Slovakia and Slovenia](https://nextstrain.org/ncov/europe/2020-04-29?d=tree,map&f_country=Hungary,Latvia,Slovakia,Slovenia&f_region=Europe&p=full)

In March, 20 sequences were submitted from Hungary, Latvia, Slovakia, and Slovenia.
While we don't yet have enough data from these countries to draw firm conclusions, we applaud these early efforts and will continue to monitor the incoming data.
<!-- There is no right-side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text 13-->
# [What you can do](https://nextstrain.org/ncov/2020-03-27?c=country&d=map&p=full)
#### ...as an individual
* Practice strict social distancing, especially if you are in a vulnerable group.
* Remember that even if you are not super vulnerable, many people around you are; follow these practices to protect others.
* Wash your hands "like you just chopped a jalapeno and have to change a contact lens."  
* Stay home as much as possible -- especially  if you are sick; be prepared with extra supplies in case you need to self-quarantine.  
* If you are an employer, encourage your employees to work from home wherever possible.

#### ...as an official  
* Make testing free and broadly available.  
* Put strong social distancing measures in place.  
* Fund and implement extensive contact tracing efforts.  
* Financially support those impacted by social distancing measures.


<!-- This is the right-side text -->

```auspiceMainDisplayMarkdown
# Takeaways

####
<br><br>
####
<br><br>



```

<!-- ############ SLIDE BREAK ############# -->






<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [Scientific credit](https://nextstrain.org/ncov/2020-XX-XX?d=map&c=author)

We would like to acknowledge the amazing and timely work done by all scientists involved in this outbreak, and particularly those working in China.
Only through the rapid sharing of genomic data and metadata are analyses such as these possible.

<br>

We also gratefully acknowledge [GISAID](https://gisaid.org) for providing the platform through which these data can be uploaded and shared.

<!-- Do not need to translate institutions names -->
<!-- This is right-side text -->
```auspiceMainDisplayMarkdown

We are grateful for the data gathered by these originating labs:



```


<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [Detailed scientific credit](https://nextstrain.org/ncov/2020-03-05?d=map&c=author)

These data were shared via [GISAID](https://gisaid.org).
We gratefully acknowledge their contributions.

<br>

To the right we give specific sequences shared by each lab.

<!-- This is right-side text -->
```auspiceMainDisplayMarkdown

The SARS-CoV-2 genomes were generously shared by scientists at these submitting labs:




```
