---
title: Genomic analysis of COVID-19 spread. Situation report 2020-05-08.
authors:
  - Cassia Wagner
  - Sidney M. Bell
  - Nicola MÃ¼ller
  - Emma Hodcroft
  - James Hadfield
  - Richard Neher
  - Trevor Bedford
authorLinks:
  - https://bedford.io/team/cassia-wagner/
  - https://twitter.com/sidneymbell
  - https://bedford.io/team/nicola-mueller/
  - https://neherlab.org/emma-hodcroft.html
  - https://bedford.io/team/james-hadfield/
  - https://neherlab.org/richard-neher.html
  - https://bedford.io/team/trevor-bedford/
affiliations: "Fred Hutch, Seattle, USA; Biozentrum, Basel, Switzerland; CZI, CA, USA"
translators:
translatorLinks:
license: "CC-BY"  
licenseLink: "https://creativecommons.org/licenses/by/4.0/"
dataset: "https://nextstrain.org/ncov/asia/2020-05-08?d=map&legend=closed&f_region=Asia"

abstract: "This weekly report uses publicly shared genomic data to track the spread of COVID-19. This week, we focus on Asia. We report many separate, independent introductions to India; context for the outbreak among migrant workers in Singapore; and the prevalence of international spread across the region."
---
<!-- Translators: Only text after : in the above ^ needs to be translated -->
<!-- Comment tags like these do not need to be translated, they are only to help you! -->
<!-- Ensure that links always end in a 'letter' (. counts) If some kind of text doesn't follow them, it breaks the slide. -->
<!-- numbers can be tagged ilke this: 161</tag> - this is just for us to help find them to update! Just leave in the </tag> bit. -->

<!-- This is left-side text 1-->
# [Table of Contents](https://nextstrain.org/ncov/2020-05-08?d=tree,map&p=grid)

* [Background resources](https://nextstrain.org/narratives/ncov/sit-rep/2020-05-08?n=2).     
* [About this data](https://nextstrain.org/narratives/ncov/sit-rep/2020-05-08?n=3).  
* [Origins of the epidemic in India](https://nextstrain.org/narratives/ncov/sit-rep/2020-05-08?n=4).  
* [Contextualizing the resurgence of cases in Singapore](https://nextstrain.org/narratives/ncov/sit-rep/2020-05-08?n=5).  
* [Introductions to Malaysia](https://nextstrain.org/narratives/ncov/sit-rep/2020-05-08?n=6).  
* [Frequent spread across international borders](https://nextstrain.org/narratives/ncov/sit-rep/2020-05-08?n=7).    
* [Scientific credit](https://nextstrain.org/narratives/ncov/sit-rep/2020-05-08?n=8).

<!-- This is right-side text -->
```auspiceMainDisplayMarkdown
# Executive summary

We analyzed 2,226 publicly shared COVID-19 genomes. By comparing these viral genomes to each other, we can characterize how COVID-19 is moving around the world and spreading locally. In this week's updates, we focus on recent spread across Asia. We report:

* Multiple introductions to Delhi during March 2020 from all over the world, rather than a single source  
* Many separate introductions to Singapore, most of which do not appear to have spread further  
* Expansion of a large cluster of cases in Singapore likely driven by sociological factors
* Similar patterns of international spread that emphasize the urgent need for cooperative efforts
```


<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text 2-->
# [COVID-19 Resources](https://nextstrain.org/ncov/asia/2020-05-08?d=tree&p=full&legend=closed&f_region=Asia)
We've prepared some resources that will make interpreting the data we present in this narrative easier.
#### Nextstrain Resources  
* [START HERE: How to read a phylogeny](https://nextstrain.org/narratives/trees-background/).  
* [Background on coronaviruses](https://nextstrain.org/help/coronavirus/human-CoV).
* [Common misconceptions](https://nextstrain.org/narratives/ncov/sit-rep/2020-03-13?n=11).

#### External Resources  
* [Ask a Scientist & FAQs](https://covid19.fas.org/l/en).
* [WHO Situation Reports](https://www.who.int/emergencies/diseases/novel-coronavirus-2019/situation-reports).
* [CDC Resources](https://www.cdc.gov/coronavirus/2019-ncov/index.html).
* [The NYTimes COVID-19 coverage](https://www.nytimes.com/news-event/coronavirus).
<!-- There is no right-side text -->


<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [A note on sampling](https://nextstrain.org/ncov/asia/2020-05-08?c=country&r=country&d=map&p=grid&legend=closed)
We currently have sequences from samples taken in 28 countries in Asia. This is an incredible achievement -- sequencing an unknown, large RNA virus in the midst of a pandemic is difficult, and is only possible through the incredible work and timely sharing of data by scientists and physicians around the world.
<br><br>
While this data enables us to infer many useful characteristics of the outbreak and track its spread in real time, it's important to emphasize that our conclusions are limited by the available data.
<br><br>
For example, the map shows very few sequences from the Middle East. This is NOT because COVID-19 isn't circulating in these areas, or that these cases are not as crucial to understand; rather, we just don't have much data available from these areas. The size of each circle on the map indicates how much data is currently available from that area, rather than the true size of the outbreak.

<!-- There is NO right-side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text 4-->
# [Many separate introductions to Delhi during March](https://nextstrain.org/ncov/asia/2020-05-06?c=division&d=tree,map&dmax=2020-04-05&dmin=2020-03-01&f_country=India&f_division=Delhi&p=full&r=country)

A [series of religious gatherings in Delhi](https://en.wikipedia.org/wiki/2020_Tablighi_Jamaat_coronavirus_hotspot_in_Delhi), between March 1-24, have been scrutinized for their presumed contribution to the spread of SARS-CoV-2 across the country.
<br><br>
Between March 1-April 8 (two weeks after the gathering), 16 viral genomes from the Delhi area were sampled and sequenced. As shown here, these cases are scattered across the tree.
<br><br>
If these gatherings were the primary hub of transmission in Delhi, we might have expected to see most of these cases grouping together in large clusters. This would have been consistent with a few infected attendees spreading the virus to many other people.
<br><br>
Instead, we see a large number of small clusters, scattered across the tree. This means that these sampled cases from Delhi were the result of many different introductions from all over the world.
<br><br>
We don't yet have enough data to directly assess the impact of this specific gathering. However, these observations provide strong evidence for many separate introductions to Delhi during this time period; this gathering is only part of the story.
<!-- There is no right side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text 5-->
# [Many cases in India are not necessarily linked to Delhi](https://nextstrain.org/ncov/asia/2020-05-06?c=division&d=tree&f_country=India&f_division=Delhi,Karnataka&label=clade:A2&m=div&p=full&r=division)

We see similar patterns of multiple independent introductions to other states across India. For example, here we see samples from Karnataka and from Delhi. The samples from Karnataka form two separate clusters, but each of these is more closely related to samples from Europe than to samples from Delhi.
<br><br>
We can't preclude the possibility that there were intermediate cases in other locations that simply weren't sampled. However, this reinforces that the outbreak in India was sparked by many separate introductions from other parts of the world.
<!-- There is no right side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text 6-->
# [Contextualizing the SARS-CoV-2 resurgence in Singapore](https://nextstrain.org/ncov/asia/2020-05-06?d=tree,map&dmin=2020-02-26&f_country=China,Singapore,India,Philippines&p=full&legend=closed&r=division)

In the last 10 weeks, SARS-CoV-2 has been introduced to Singapore many times, from multiple locations.
<br><br>
Most of these introductions have led to limited onward transmission (single cases or small clusters on the tree).
<br><br>
In the middle of the tree, however, we see a single introduction that led to extensive onward transmission.
This large, cohesive cluster is most closely related to samples from India and the Philippines.
<br><br>
This is consistent with [recent reports](https://www.nytimes.com/interactive/2020/04/28/world/asia/coronavirus-singapore-migrants.html) that the virus is spreading rapidly among vulnerable, confined populations of migrant workers in Singapore.
<br><br>
This illustrates how the extent to which an introduction spreads within a population depends on sociological factors and emphasizes why it is critically important to protect vulnerable groups.
<!-- There is no right side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Repeated introductions to Malaysia](https://nextstrain.org/ncov/asia/2020-05-08?f_country=Malaysia&d=tree,map&r=division)

We have 14 sequences from Malaysia on Nextstrain, sampled between late January and late March. These sequences group with several different clades, each of which was circulating mostly in other Asian countries. This suggests that there have been multiple introductions from the surrounding region into Malaysia.


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Regional spread across Asia](https://nextstrain.org/ncov/asia/2020-05-08?d=tree,map&f_country=Japan,Jordan,Thailand,United%20Arab%20Emirates)

A similar pattern is repeated for many other countries in the region: multiple introductions, many of which were well-contained, but with a few that took off to spark local transmission. We encourage you to "Explore the Data Yourself". You will observe similar patterns in Jordan, Japan, Thailand, and the United Arab Emirates.
<br><br>
The consistency of this pattern is strong evidence that the virus frequently crosses international borders. Thus, in order to effectively control the epidemic in any one country, we must control the pandemic across all countries through coordinated international efforts.
<!-- There is no right side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text 13-->
# [What you can do](https://nextstrain.org/ncov/2020-05-08?c=country&d=map&p=full)
#### ...as an individual
* Practice strict social distancing, especially if you are in a vulnerable group.
* Remember that even if you are not super vulnerable, many people around you are; follow these practices to protect others.
* Wash your hands "like you just chopped a jalapeno and have to change a contact lens."  
* Stay home as much as possible -- especially  if you are sick; be prepared with extra supplies in case you need to self-quarantine.  
* If you are an employer, encourage your employees to work from home wherever possible.

#### ...as an official  
* Make testing free and broadly available.  
* Put strong social distancing measures in place.  
* Fund and implement extensive contact tracing efforts.  
* Financially support those impacted by social distancing measures.


<!-- This is the right-side text -->

```auspiceMainDisplayMarkdown
# Takeaways

#### The outbreak in India is the result of many separate introductions from around the world, rather than a single gathering.  
<br><br>
#### In Singapore, sociological factors accelerated the expansion of one introduction to become a large outbreak, in contrast to the many introductions which have not spread widely.
<br><br>
#### In Asia, most countries' outbreaks have a similar history: many separate introductions from other countries, some of which resulted in local spread. This pattern of multiple introductions over time emphasizes the need for international and regional cooperation to fully control the pandemic.
```

<!-- ############ SLIDE BREAK ############# -->






<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [Scientific credit](https://nextstrain.org/ncov/2020-05-06?d=map&c=author)

We would like to acknowledge the amazing and timely work done by all scientists involved in this outbreak.
Only through the rapid sharing of genomic data and metadata are analyses such as these possible.
<br><br>
The specific sequences provided by each lab are [listed here]().
<br><br>
We also gratefully acknowledge GISAID for providing the platform through which these data can be uploaded and shared.

<!-- Do not need to translate institutions names -->
<!-- This is right-side text -->
```auspiceMainDisplayMarkdown

We are grateful for the data gathered by these originating labs:



```
