---
title: Genomic analysis of COVID-19 spread. Situation report 2020-04-24.
authors:
  - Cassia Wagner
  - Sidney M. Bell
  - Emma Hodcroft
  - Nicola Müller
  - James Hadfield
  - Misja Ilcisin
  - Richard Neher
  - Trevor Bedford
authorLinks:
  - https://bedford.io/team/cassia-wagner/
  - https://twitter.com/sidneymbell
  - https://neherlab.org/emma-hodcroft.html
  - https://bedford.io/team/nicola-mueller/
  - https://bedford.io/team/james-hadfield/
  - https://bedford.io/team/misja-ilcisin/
  - https://neherlab.org/richard-neher.html
  - https://bedford.io/team/trevor-bedford/
affiliations: "Fred Hutch, Seattle, USA; Biozentrum, Basel, Switzerland; CZI, CA, USA"
translators:
translatorLinks:
license: "CC-BY"  
licenseLink: "https://creativecommons.org/licenses/by/4.0/"
dataset: "https://nextstrain.org/ncov/africa/2020-04-23?f_region=Africa&d=map,tree"

abstract: "This weekly report uses publicly shared genomic data to track the spread of COVID-19. This week, we focus on Africa. We report many separate introductions to the Democratic Republic of the Congo, Senegal, Ghana, Nigeria, Gambia, Algeria, and South Africa. We find evidence of local transmission in Kinshasa, DRC; Dakar and Toubeh, Senegal; and Greater Accra, Ghana."
---
<!-- Translators: Only text after : in the above ^ needs to be translated -->
<!-- Comment tags like these do not need to be translated, they are only to help you! -->
<!-- Ensure that links always end in a 'letter' (. counts) If some kind of text doesn't follow them, it breaks the slide. -->
<!-- numbers can be tagged ilke this: 161</tag> - this is just for us to help find them to update! Just leave in the </tag> bit. -->

<!-- This is left-side text -->
# [Table of Contents](https://nextstrain.org/ncov/africa/2020-04-23?d=map,tree&f_region=Africa)

* [Background resources](https://nextstrain.org/narratives/ncov/sit-rep/2020-04-24?n=2).     
* [About this data](https://nextstrain.org/narratives/ncov/sit-rep/2020-04-24?n=3).
* [The genetic diversity of SARS-CoV-2 in Africa](https://nextstrain.org/narratives/ncov/sit-rep/2020-04-24?n=X).
* [Updates for the DRC](https://nextstrain.org/narratives/ncov/sit-rep/2020-04-24?n=X).
* [Updates for Senegal](https://nextstrain.org/narratives/ncov/sit-rep/2020-04-24?n=X).
* [Updates for Ghana](https://nextstrain.org/narratives/ncov/sit-rep/2020-04-24?n=X).
* [Updates for Algeria, Gambia, Nigeria, and South Africa](https://nextstrain.org/narratives/ncov/sit-rep/2020-04-24?n=X).
* [What you can do](https://nextstrain.org/narratives/ncov/sit-rep/2020-04-24?n=X).
* [FAQ & common misconceptions](https://nextstrain.org/narratives/ncov/sit-rep/2020-04-24?n=X).
* [Scientific credit](https://nextstrain.org/narratives/ncov/sit-rep/2020-04-24?n=X).

<!-- This is right-side text -->
```auspiceMainDisplayMarkdown
# Executive summary

We analyzed 1,300 publicly shared COVID-19 genomes. By comparing these viral genomes to each other, we can characterize how COVID-19 is moving around the world and spreading locally. In this week's updates, we report:
<br><br>
* Many separate introductions of SARS-CoV-2 to the African continent, primarily from Europe (although sampling bias may contribute to this inference)
<br><br>
* At least 7 introductions to the DRC  
<br><br>
* 2 separate transmission chains circulating in Kinshasa, DRC  
<br><br>
* Of 8 introductions to Senegal, at least 2 seeded local transmission chains in Dakar and Touba
<br><br>
* Local transmission in Greater Accra, Ghana, after 7 introductions to this region
<br><br>
* Scattered introductions to Algeria, Gambia, Nigeria, and South Africa
```


<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text 2-->
# [COVID-19 Resources](https://nextstrain.org/ncov/africa/2020-04-23?d=tree&p=full&legend=closed&f_region=Africa)
We've prepared some resources that will make interpreting the data we present in this narrative easier.
#### Nextstrain Resources  
* [START HERE: How to read a phylogeny](https://nextstrain.org/narratives/trees-background/).  
* [Background on coronaviruses](https://nextstrain.org/help/coronavirus/human-CoV).
* [Common misconceptions](https://nextstrain.org/narratives/ncov/sit-rep/2020-03-13?n=11).

#### External Resources  
* [Ask a Scientist & FAQs](https://covid19.fas.org/l/en).
* [WHO Situation Reports](https://www.who.int/emergencies/diseases/novel-coronavirus-2019/situation-reports).
* [CDC Resources](https://www.cdc.gov/coronavirus/2019-ncov/index.html).
* [The NYTimes COVID-19 coverage](https://www.nytimes.com/news-event/coronavirus).
<!-- There is no right-side text -->


<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [A note on sampling](https://nextstrain.org/ncov/africa/2020-04-23?d=map&f_region=Africa&p=full)
We currently have sequences from samples taken in 6 continents, including 7 African countries. This is an incredible achievement -- sequencing an unknown, large RNA virus in the midst of a pandemic is difficult, and is only possible through the incredible work and timely sharing of data by scientists and physicians around the world.
<br><br>
While this data enables us to infer many useful characteristics of the outbreak and track its spread in real time, it's important to emphasize that our conclusions are limited by the available data.
<br><br>
For example, the map shows very few sequences from East Africa. This is NOT because COVID-19 isn't circulating in these areas, or that these cases are not as crucial to understand; rather, we just don't have much data available from these areas. The size of each circle on the map indicates how much data is currently available from that area, rather than the true size of the outbreak.

<!-- There is NO right-side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Many separate introductions have seeded diverse local transmission chains in Africa](https://nextstrain.org/ncov/africa/2020-04-23?d=tree,map&f_region=Africa&p=full&r=country&legend=open)
  At the time of writing, the [Africa CDC](https://africacdc.org/covid-19/) reports 26,144 confirmed COVID-19 cases and 1,247 deaths across the African Union.
  We have viral genomes from 83 of these cases from 7 African nations, shown here (in color) against a backdrop of contextual sequences from around the world (gray). We applaud the local public health departments for their rapid response and open data sharing.
  <br><br>
  These samples span the full breadth of the tree, indicating that there have been many independent introductions to the African continent.
  Many of these introductions appear to be from Europe, although this may be skewed by the fact that we have more sequences from Europe than from most other places in the world; we can't rule out the possibility of unsampled, intermediate transmissions elsewhere.
  <br><br>
  Not all of these introductions have led to community transmission, but we see clear clusters in some locations.

<!-- There is no right side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [At least 7 introductions to the DRC](https://nextstrain.org/ncov/africa/2020-04-23?d=tree,map&f_region=Africa&p=full&r=country&legend=closed&f_country=Democratic%20Republic%20of%20the%20Congo)

There are 359 confirmed cases of COVID-19 in the DRC with 25 deaths. On Nextstrain, we have 36 sequences from the DRC.
<br><br>
These sequences fall across the tree, with evidence for at least 7 separate introductions. Heuristically, introductions look like a color change between a virus sample (tips of the tree) and its ancestors (the internal nodes or branch points that lead to it). From these 7 introductions, we see evidence for at least 2 local transmission chains.

<!-- There is no right side text -->

<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [2 local transmission chains in Kinshasa, seeded in early March](https://nextstrain.org/ncov/africa/2020-04-23?d=tree&f_country=Democratic%20Republic%20of%20the%20Congo&f_region=Africa&label=clade:A2a&p=full&legend=closed&m=div)

In this part of the tree, we see two clusters representing likely local transmission.
Local transmission looks like a tight cluster of cases from the same location, sampled over time.
<br><br>
Towards the top of this view, we see a cluster of cases from Kinshasa, all sampled between March 9 and March 22.
<br><br>
Towards the middle, we see another cluster of cases from Kinshasa, sampled between March 15 and March 18th.
<br><br>
From this, we can conclude that there were at least 2 separate transmission chains circulating in Kinshasa in early March, 2020.
<br><br>
The DRC reported it’s first cases of COVID-19 to the WHO on March 10. The first sample on Nextstrain was collected on March 9, and was reported in our [Situation Report for March 27](https://nextstrain.org/narratives/ncov/sit-rep/2020-03-27?n=16).
<!-- There is no right-side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [At least 8 introductions to Senegal, with 2 local transmission chains](https://nextstrain.org/ncov/africa/2020-04-23?d=tree,map&f_country=Senegal&f_region=Africa&m=div&p=full)

Similar to the patterns found in the DRC, here we see samples from Senegal spanning the full genetic diversity of the tree. The phylogenetic placement of these samples indicates that there were at least 8 separate introductions to Senegal between late February and late March. Senegal reported its first  COVID-19 cases to the WHO on February 28, the same day that the first viral sample was collected for sequencing.
<br><br>
As we reported in our [April 3 Situation Report](https://nextstrain.org/narratives/ncov/sit-rep/2020-04-03?n=5), at least 2 of these introductions in late February or early March have led to ongoing local transmission clusters in Dakar and Tabou, respectively.


<!-- There is no right-side text --><!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [At least 7 introductions to Ghana](https://nextstrain.org/ncov/africa/2020-04-23?d=tree,map&f_country=Ghana&f_region=Africa&p=full)

There are 1,154 confirmed cases of COVID-19 in Ghana with 9 known deaths. On Nextstrain, there are 14 sequences from Ghana which are spread across the phylogenetic tree, representing at least 7 separate introductions.
<br><br>
The first case in Ghana was reported to the WHO on March 11, and the first viral genome on Nextstrain was collected on March 24.

<!-- There is no right-side text --><!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Local transmission in Greater Accra, Ghana](https://nextstrain.org/ncov/africa/2020-04-23?label=clade:B4&c=division&d=tree,map&f_country=Ghana,Senegal&f_region=Africa&p=full&r=division&legend=closed)

Here, we see a cluster of closely related cases from the Greater Accra region of Ghana, sampled between March 25 and March 30.
The common ancestor of these cases dates to mid-February or early March, and was most likely imported from Asia.
<br><br>
Interestingly, there is a sample from Senegal that groups tightly with this cluster, suggesting the possibility of a transmission between Senegal and Ghana.

<!-- There is no right-side text --><!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Scattered introductions to Algeria, Gambia, Nigeria, and South Africa](https://nextstrain.org/ncov/africa/2020-04-23?c=division&d=tree,map&f_country=Algeria,Gambia,Nigeria,South%20Africa&f_region=Africa&p=full)

Here, we see samples isolated from Algeria, Gambia, Nigeria and South Africa. These samples are scatted across the full tree, and are largely isolated from one another.
<br><br>
This is evidence for many separate introductions, largely from Europe, to many nations in Africa. Without more data, we can't yet say which -- if any -- of these sparks resulted into local outbreaks.


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text 13-->
# [What you can do](https://nextstrain.org/ncov/2020-03-27?c=country&d=map&p=full)
#### ...as an individual
* Practice strict social distancing, especially if you are in a vulnerable group.
* Remember that even if you are not super vulnerable, many people around you are; follow these practices to protect others.
* Wash your hands "like you just chopped a jalapeno and have to change a contact lens."  
* Stay home as much as possible -- especially  if you are sick; be prepared with extra supplies in case you need to self-quarantine.  
* If you are an employer, encourage your employees to work from home wherever possible.

#### ...as an official  
* Make testing free and broadly available.  
* Put strong social distancing measures in place.  
* Fund and implement extensive contact tracing efforts.  
* Financially support those impacted by social distancing measures.


<!-- This is the right-side text -->

```auspiceMainDisplayMarkdown
# Takeaways

#### There have been many separate introductions of SARS-CoV-2 to the African continent, primarily from Europe (although sampling bias may contribute to this inference).
<br><br>
#### 7 introductions to the DRC have led to at least 2 separate transmission chains circulating in Kinshasa.  
<br><br>
#### Similarly, 8 introductions to Senegal seeded local transmission chains in Dakar and Touba.  
<br><br>
#### SARS-CoV-2 has been introduced to Ghana at least 7 times, with local transmission in Greater Accra.
<br><br>
#### There have been scattered introductions to Algeria, Gambia, Nigeria, and South Africa. We don't yet have enough samples to infer local transmission from the sequence data.
```

<!-- ############ SLIDE BREAK ############# -->






<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [Scientific credit](https://nextstrain.org/ncov/2020-04-24?d=map&c=author)

We would like to acknowledge the amazing and timely work done by all scientists involved in this outbreak, and particularly those working in China.
Only through the rapid sharing of genomic data and metadata are analyses such as these possible.

<br>

We also gratefully acknowledge [GISAID](https://gisaid.org) for providing the platform through which these data can be uploaded and shared.

<!-- Do not need to translate institutions names -->
<!-- This is right-side text -->
```auspiceMainDisplayMarkdown

We are grateful for the data gathered by these originating labs:



```


<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [Detailed scientific credit](https://nextstrain.org/ncov/2020-03-05?d=map&c=author)

These data were shared via [GISAID](https://gisaid.org).
We gratefully acknowledge their contributions.

<br>

To the right we give specific sequences shared by each lab.

<!-- This is right-side text -->
```auspiceMainDisplayMarkdown

The SARS-CoV-2 genomes were generously shared by scientists at these submitting labs:




```
