---
title: Genomic analysis of COVID-19 spread. Situation report 2020-04-10.
authors:
  - Sidney M. Bell
  - Nicola MÃ¼ller
  - Cassia Wagner
  - Emma Hodcroft
  - James Hadfield
  - Richard Neher
  - Trevor Bedford
authorLinks:
  - https://twitter.com/sidneymbell
  - https://bedford.io/team/nicola-mueller/
  - https://bedford.io/team/cassia-wagner/
  - https://neherlab.org/emma-hodcroft.html
  - https://bedford.io/team/james-hadfield/
  - https://neherlab.org/richard-neher.html
  - https://bedford.io/team/trevor-bedford/
affiliations: "Fred Hutch, Seattle, USA; Biozentrum, Basel, Switzerland; CZI, CA, USA"
translators:
translatorLinks:
date: "2020 April 9"
dataset: "https://nextstrain.org/ncov/global/2020-04-09?d=tree,map&p=full&legend=closed"
abstract: "This report uses publicly shared genomic data to track the spread of COVID-19. These reports are updated weekly."
---
<!-- Translators: Only text after : in the above ^ needs to be translated -->
<!-- Comment tags like these do not need to be translated, they are only to help you! -->
<!-- Ensure that links always end in a 'letter' (. counts) If some kind of text doesn't follow them, it breaks the slide. -->
<!-- numbers can be tagged ilke this: 161</tag> - this is just for us to help find them to update! Just leave in the </tag> bit. -->

<!-- This is left-side text -->
*/ # [Table of Contents](https://nextstrain.org/ncov/2020-XX-XX?d=tree,map&p=grid)

* [Background resources](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=2).     
* [About this data](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=3).
* [](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=X).

* [What you can do](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=X).
* [FAQ & common misconceptions](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=X).
* [Scientific credit](https://nextstrain.org/narratives/ncov/sit-rep/2020-XX-XX?n=X). /*

<!-- This is right-side text -->
```auspiceMainDisplayMarkdown
# Executive summary

We analyzed 3,160 publicly shared SARS-Cov-2 genomes. By comparing these viral genomes to each other, we can characterize how SARS-CoV-2 (which causes COVID-19) is moving around the world and spreading locally.


We are in the process of moving to more specific, regionally-focused analyses. You can find more information about this change [here]() and how to help us out [here](); thank you for your patience during this transition.


This week's report focuses on transmission dynamics _between_ regions of the world. Here, we cover:  

* Early spread within Asia    
* Initial introductions from Asia to North America, Europe and Oceania  
* Growth of North American & European outbreaks  
* Subsequent spread across international borders within regions  
* Spread from Europe and North America to the global south  
* Coming full circle: reintroductions to Asia  
```


<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [COVID-19 Resources](https://nextstrain.org/ncov/2020-03-27)
We've prepared some resources that are worth reading to familiarize yourself with COVID-19 and the virus that causes it, SARS-CoV-2, also known as hCov-19.
This information will make interpreting the data we present in this narrative easier.
<br><br>
If you aren't familiar with phylogenetic trees, we encourage you to check out the ['How to Read Phylogenies' narrative](https://nextstrain.org/narratives/trees-background/) and come back when you're ready.
<!-- This is right-side text -->
```auspiceMainDisplayMarkdown

## Nextstrain Resources  

[How to read a phylogeny](https://nextstrain.org/narratives/trees-background/).  

[Background on coronaviruses](https://nextstrain.org/help/coronavirus/human-CoV).

[FAQs](https://nextstrain.org/help/coronavirus/FAQ).

[Common misconceptions](https://nextstrain.org/narratives/ncov/sit-rep/2020-03-13?n=11).


## External Resources  
[WHO Situation Reports](https://www.who.int/emergencies/diseases/novel-coronavirus-2019/situation-reports).

[CDC Resources](https://www.cdc.gov/coronavirus/2019-ncov/index.html).

[The NYTimes COVID-19 coverage](https://www.nytimes.com/news-event/coronavirus).

[Our World in Data](https://ourworldindata.org/coronavirus).

[Ask a Scientist & FAQs](https://covid19.fas.org/l/en).


```

<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [A note on sampling](https://nextstrain.org/ncov/global/2020-04-10?c=country&r=country&d=map&p=grid&legend=closed)
We currently have sequences from samples taken in 57 countries across 6 continents. This is an incredible achievement -- sequencing an unknown, large RNA virus in the midst of a pandemic is difficult, and is only possible through the incredible work and timely sharing of data by scientists and physicians around the world.
<br><br>
While this data enables us to infer many useful characteristics of the outbreak and track its spread in real time, it's important to emphasize that our conclusions are limited by the available data.
<br><br>
For example, the map shows very few sequences from the global south. This is NOT because COVID-19 isn't circulating in these areas, or that these cases are not as crucial to understand; rather, we just don't have much data available from these areas. The size of each circle on the map indicates how much data is currently available from that area, rather than the true size of the outbreak.

<!-- There is NO right-side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Early spread within Asia](https://nextstrain.org/ncov/global/2020-04-09?d=tree,map&dmax=2020-01-15&p=full)

Here, we highlight samples from mid-December 2019 through mid-January 2020, during the first month of the outbreak.
Tips of the tree represent samples from infected individuals.
Each internal node, or branch point, represents an unsampled case, which was the ancestor of all of the other cases within the descendant clade.
<br><br>
Tips are colored according to where they were sampled from; nodes (branch points) are colored by the inferred (most likely) location of the ancestral case.
<br><br>
Comparing where a sequence was isolated from geographically to where its ancestors were circulating can help us understand not only where infections are occurring, but also where they may have been acquired.
<br><br>
The common ancestor of circulating viruses appears to have emerged in Wuhan, China, in late Nov or early Dec 2019.
Accordingly, the majority of sampled and inferred ancestral cases were located in Asia during this early period.
This is consistent with circulation largely within China and some exported cases to other countries in Asia for the first month of the outbreak.  

<!-- There is no right side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Initial introductions to other regions](https://nextstrain.org/ncov/global/2020-04-09?dmax=2020-02-15&dmin=2020-01-15&p=full&d=map,tree)

During the 2nd month of the outbreak, from mid-January to mid_February 2020, we start to see individual cases crop up in North America, Europe and Oceania.
These early samples still group with samples from Asia, and generally lack the hallmarks of local transmission.
<br><br>
This is consistent with initial sparks being thrown from the main outbreak in Asia to other parts of the world. Although most of these sparks did not result in localized outbreaks, at least two of these introductions took off.
<!-- There is no right side text -->

<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Cryptic transmission in Europe from late January](https://nextstrain.org/ncov/global/2020-04-09?d=tree,map&dmax=2020-02-14&label=clade:A2&p=full)

Zooming in to the top of the tree for the same time period, we see an ancestral European virus (internal node) that seeded the large European outbreak.
<br><br>
This indicates that while we hadn't yet sequenced any samples from this clade, the origins of the European outbreak were already circulating undetected by late January.

<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Cryptic transmission in North America from late January](https://nextstrain.org/ncov/global/2020-04-09?d=tree,map&dmax=2020-02-14&label=clade:B1&p=full)

Similarly, zooming in to the bottom of the tree, we see an ancestral North American virus (internal node) that initiated what would later become a large outbreak in North America.
<br><br>
This indicates that although we only had 1 sequenced case from this cluster at the time, the virus was circulating undetected within North America by late January.

<!-- There is no right side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Expansion of North American and European outbreaks](https://nextstrain.org/ncov/global/2020-04-09?dmax=2020-03-15&dmin=2020-02-15&p=full&d=tree,map)

During the 3rd month of the outbreak, from mid-February to mid-March 2020, these two clusters expanded to cause large outbreaks within Europe and North America, respectively.
<!-- There is no right side text -->


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Spread across international borders](https://nextstrain.org/ncov/global/2020-04-09?dmax=2020-03-15&dmin=2020-02-15&f_region=North%20America&p=full&r=division&d=tree,map&c=division)

Within each of these regions, we see widespread mixing across state lines and international borders.
This is demonstrated here by the intermingling of samples from multiple U.S. states and Canadian provinces.
<br><br>
We've [previously reported](https://nextstrain.org/narratives/ncov/sit-rep/2020-03-27?f_region=Europe&n=5) on a similar pattern of frequent mixing of viral lineages between European countries.
<!-- There is no right side text -->

<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Introductions from North America & Europe to other parts of the globe](https://nextstrain.org/ncov/global/2020-04-09?dmax=2020-03-15&dmin=2020-02-15&f_region=Africa,South%20America,Oceania&p=full&d=tree,map)

These clusters are not restricted to their respective regions. During this time period, we also see cases from South America, Oceania, and Africa against a backdrop of primarily European and North American clades. This indicates that there were frequent intercontinental introductions during this time period.
<br><br>
Interestingly, samples from the global south span the full breadth of the tree, indicating that these regions' outbreaks are mixtures of many viral lineages.


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text -->
# [Coming full circle: reintroductions to Asia](https://nextstrain.org/ncov/global/2020-04-09?d=tree,map&dmin=2020-03-14&f_region=Asia&p=full&r=division)

In the last month, the pandemic has come full circle. Here, we see re-introductions from Europe and North America back to Asia.
<br><br>
This resoundingly demonstrates that this pandemic is a global fight; we all must help control this virus everywhere if we hope to defeat it anywhere.


<!-- ############ SLIDE BREAK ############# -->
<!-- This is left-side text 13-->
# [Takeaways](https://nextstrain.org/ncov/2020-03-27?c=country&d=map&p=full)

<!-- This is the right-side text -->

```auspiceMainDisplayMarkdown
# [What you can do](https://nextstrain.org/ncov/2020-03-27?c=country&d=map&p=full)
#### ...as an individual
* Practice strict social distancing, especially if you are in a vulnerable group.
* Remember that even if you are not super vulnerable, many people around you are; follow these practices to protect others.
* Wash your hands "like you just chopped a jalapeno and have to change a contact lens."  
* Stay home as much as possible -- especially  if you are sick; be prepared with extra supplies in case you need to self-quarantine.  
* If you are an employer, encourage your employees to work from home wherever possible.

#### ...as an official  
* Make testing free and broadly available.  
* Put strong social distancing measures in place.  
* Fund and implement extensive contact tracing efforts.  
* Financially support those impacted by social distancing measures.

```

<!-- ############ SLIDE BREAK ############# -->






<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [Scientific credit](https://nextstrain.org/ncov/2020-04-10?d=map&c=author)

We would like to acknowledge the amazing and timely work done by all scientists involved in this outbreak, and particularly those working in China.
Only through the rapid sharing of genomic data and metadata are analyses such as these possible.

<br>

We also gratefully acknowledge [GISAID](https://gisaid.org) for providing the platform through which these data can be uploaded and shared.

<!-- Do not need to translate institutions names -->
<!-- This is right-side text -->
```auspiceMainDisplayMarkdown

We are grateful for the data gathered by these originating labs:



```


<!-- ############ SLIDE BREAK ############# -->

<!-- This is left-side text -->
# [Detailed scientific credit](https://nextstrain.org/ncov/2020-03-05?d=map&c=author)

These data were shared via [GISAID](https://gisaid.org).
We gratefully acknowledge their contributions.

<br>

To the right we give specific sequences shared by each lab.

<!-- This is right-side text -->
```auspiceMainDisplayMarkdown

The SARS-CoV-2 genomes were generously shared by scientists at these submitting labs:




```
